// @sanity/runtime-cli/dist/actions/blueprints/logs.js
import config from '../../config.js';
import getHeaders from '../../utils/get-headers.js';
import { createTracedFetch } from '../../utils/traced-fetch.js';
const { apiUrl } = config;
export const logsUrl = `${apiUrl}vX/blueprints/logs`;
export async function getLogs(params, auth, logger) {
    const fetchFn = createTracedFetch(logger);
    const url = new URL(logsUrl);
    if (params.stackId)
        url.searchParams.append('stackId', params.stackId);
    if (params.operationId)
        url.searchParams.append('operationId', params.operationId);
    if (params.limit)
        url.searchParams.append('limit', String(params.limit));
    if (params.before)
        url.searchParams.append('before', params.before);
    const response = await fetchFn(url.toString(), {
        headers: getHeaders(auth),
        method: 'GET',
    });
    const result = await response.json();
    return {
        ok: response.ok,
        error: response.ok ? null : result.message || 'Unknown error',
        logs: response.ok ? result : [],
    };
}
// Get most recent logs, up to a limit
export function getRecentLogs(logs, limit = 10) {
    const sortedLogs = [...logs].sort((a, b) => new Date(a.timestamp).getTime() - new Date(b.timestamp).getTime());
    return sortedLogs.slice(-limit);
}
