// @sanity/sdk/dist/index.js
import { Observable, share, map, distinctUntilChanged, skip, filter, exhaustMap, from, timer, switchMap, takeWhile, firstValueFrom, fromEvent, EMPTY, defer, asapScheduler, combineLatest, of, concatMap, withLatestFrom, concat, throwError, first as first$1, Subject, takeUntil, partition, merge, shareReplay, tap as tap$1, catchError as catchError$1, startWith as startWith$1, pairwise as pairwise$1, groupBy as groupBy$1, mergeMap as mergeMap$1, throttle, race, NEVER, Subscription, debounceTime } from "rxjs";
import { createClient, CorsOriginError } from "@sanity/client";
import { pick, omit, isEqual, isObject } from "lodash-es";
import { devtools } from "zustand/middleware";
import { createStore } from "zustand/vanilla";
import { first, switchMap as switchMap$1, groupBy, mergeMap, startWith, pairwise, filter as filter$1, map as map$1, delay, tap, catchError, scan, share as share$1 } from "rxjs/operators";
import { createController, createNode } from "@sanity/comlink";
import { createSelector } from "reselect";
import { SanityEncoder } from "@sanity/mutate";
import { getPublishedId as getPublishedId$1 } from "@sanity/client/csm";
import { jsonMatch, stringifyPath, slicePath, getIndexForKey } from "@sanity/json-match";
import { getIndexForKey as getIndexForKey2, getPathDepth, joinPaths, jsonMatch as jsonMatch2, slicePath as slicePath2, stringifyPath as stringifyPath2 } from "@sanity/json-match";
import { evaluateSync, parse } from "groq-js";
import { diffValue } from "@sanity/diff-patch";
import { applyPatches, parsePatch } from "@sanity/diff-match-patch";
import { isKeySegment, isKeyedObject } from "@sanity/types";
import { createDocumentLoaderFromClient } from "@sanity/mutate/_unstable_store";
import { SDK_CHANNEL_NAME, SDK_NODE_NAME } from "@sanity/message-protocol";
import { fromUrl } from "@sanity/bifur-client";
import { DocumentId, getPublishedId as getPublishedId$2, getDraftId as getDraftId$1, getVersionId, isDraftId, isVersionId, isPublishedId } from "@sanity/id-utils";
function isDatasetSource(source) {
  return "projectId" in source && "dataset" in source;
}
function isMediaLibrarySource(source) {
  return "mediaLibraryId" in source;
}
function isCanvasSource(source) {
  return "canvasId" in source;
}
const isReleasePerspective = (perspective) => typeof perspective == "object" && perspective !== null && "releaseName" in perspective;
function getPublishedId(id) {
  const draftsPrefix = "drafts.";
  return id.startsWith(draftsPrefix) ? id.slice(draftsPrefix.length) : id;
}
function getDraftId(id) {
  const draftsPrefix = "drafts.";
  return id.startsWith(draftsPrefix) ? id : `${draftsPrefix}${id}`;
}
function insecureRandomId() {
  return Array.from({ length: 16 }, () => Math.floor(Math.random() * 16).toString(16)).join("");
}
const LOG_LEVEL_PRIORITY = {
  error: 0,
  warn: 1,
  info: 2,
  debug: 3,
  trace: 4
}, DEFAULT_CONFIG = {
  level: "warn",
  namespaces: [],
  internal: !1,
  timestamps: !0,
  enableInProduction: !1,
  handler: {
    // eslint-disable-next-line no-console
    error: console.error.bind(console),
    // eslint-disable-next-line no-console
    warn: console.warn.bind(console),
    // eslint-disable-next-line no-console
    info: console.info.bind(console),
    // eslint-disable-next-line no-console
    debug: console.debug.bind(console),
    // eslint-disable-next-line no-console
    trace: console.debug.bind(console)
    // trace uses console.debug
  }
};
function parseDebugEnvVar() {
  if (typeof process > "u" || !process.env?.DEBUG)
    return null;
  const debug = process.env.DEBUG;
  if (!debug.includes("sanity"))
    return null;
  const config = {}, levelMatch = debug.match(/sanity:(trace|debug|info|warn|error):/), hasLevelSpecifier = !!levelMatch;
  if (levelMatch ? config.level = levelMatch[1] : config.level = "debug", debug === "sanity")
    config.namespaces = ["*"];
  else if (hasLevelSpecifier && debug.match(/sanity:(trace|debug|info|warn|error):\*/))
    config.namespaces = ["*"];
  else if (!hasLevelSpecifier && debug.includes("sanity:*"))
    config.namespaces = ["*"];
  else {
    const namespaces = debug.split(",").filter((s) => s.includes("sanity:")).map((s) => {
      const cleaned = s.replace(/^sanity:/, "");
      return hasLevelSpecifier && cleaned.match(/^(trace|debug|info|warn|error):/) ? cleaned.split(":").slice(1).join(":") : cleaned.split(":")[0];
    }).filter(Boolean).filter((ns) => ns !== "*");
    namespaces.length > 0 && (config.namespaces = namespaces);
  }
  return debug.includes(":internal") && (config.internal = !0), config;
}
const envConfig = parseDebugEnvVar();
let globalConfig = {
  ...DEFAULT_CONFIG,
  ...envConfig ?? {}
};
envConfig && (["info", "debug", "trace"].includes(globalConfig.level) || globalConfig.level === "warn") && console.info(
  `[${(/* @__PURE__ */ new Date()).toISOString()}] [INFO] [sdk] Logging auto-configured from DEBUG environment variable`,
  {
    level: globalConfig.level,
    namespaces: globalConfig.namespaces,
    internal: globalConfig.internal,
    source: "env:DEBUG",
    value: typeof process < "u" ? process.env?.DEBUG : void 0
  }
);
function configureLogging$1(config) {
  globalConfig = {
    ...globalConfig,
    ...config,
    handler: config.handler ?? globalConfig.handler
  };
}
function isLoggingEnabled() {
  return typeof process < "u" && process.env?.NODE_ENV === "production" ? globalConfig.enableInProduction : !0;
}
function isNamespaceEnabled(namespace) {
  return isLoggingEnabled() ? globalConfig.namespaces.includes("*") ? !0 : globalConfig.namespaces.includes(namespace) : !1;
}
function isLevelEnabled(level) {
  return isLoggingEnabled() ? LOG_LEVEL_PRIORITY[level] <= LOG_LEVEL_PRIORITY[globalConfig.level] : !1;
}
function formatMessage(namespace, level, message, context) {
  const parts = [];
  if (globalConfig.timestamps) {
    const timestamp = (/* @__PURE__ */ new Date()).toISOString();
    parts.push(`[${timestamp}]`);
  }
  parts.push(`[${level.toUpperCase()}]`), parts.push(`[${namespace}]`);
  const instanceContext = context?.instanceContext;
  return instanceContext && (instanceContext.projectId && parts.push(`[project:${instanceContext.projectId}]`), instanceContext.dataset && parts.push(`[dataset:${instanceContext.dataset}]`), instanceContext.instanceId && parts.push(`[instance:${instanceContext.instanceId.slice(0, 8)}]`)), parts.push(message), [parts.join(" "), context];
}
function sanitizeContext(context) {
  if (!context || Object.keys(context).length === 0) return;
  const sanitized = { ...context }, sensitiveKeys = ["token", "password", "secret", "apiKey", "authorization"];
  for (const key of Object.keys(sanitized))
    sensitiveKeys.some((sensitive) => key.toLowerCase().includes(sensitive)) && (sanitized[key] = "[REDACTED]");
  return sanitized;
}
function createLogger(namespace, baseContext) {
  const logAtLevel = (level, message, context) => {
    if (!isNamespaceEnabled(namespace) || !isLevelEnabled(level) || context?.internal && !globalConfig.internal) return;
    const mergedContext = { ...baseContext, ...context }, sanitized = sanitizeContext(mergedContext), [formatted, finalContext] = formatMessage(namespace, level, message, sanitized);
    globalConfig.handler[level](formatted, finalContext);
  };
  return {
    namespace,
    error: (message, context) => logAtLevel("error", message, context),
    warn: (message, context) => logAtLevel("warn", message, context),
    info: (message, context) => logAtLevel("info", message, context),
    debug: (message, context) => logAtLevel("debug", message, context),
    trace: (message, context) => logAtLevel("trace", message, { ...context, internal: !0 }),
    isLevelEnabled: (level) => isNamespaceEnabled(namespace) && isLevelEnabled(level),
    child: (childContext) => createLogger(namespace, { ...baseContext, ...childContext }),
    getInstanceContext: () => baseContext?.instanceContext
  };
}
function createSanityInstance(config = {}) {
  const instanceId = crypto.randomUUID(), disposeListeners = /* @__PURE__ */ new Map(), disposed = { current: !1 }, instanceContext = {
    instanceId,
    projectId: config.projectId,
    dataset: config.dataset
  }, logger = createLogger("sdk", { instanceContext });
  logger.info("Sanity instance created", {
    hasProjectId: !!config.projectId,
    hasDataset: !!config.dataset,
    hasAuth: !!config.auth,
    hasPerspective: !!config.perspective
  }), logger.debug("Instance configuration", {
    projectId: config.projectId,
    dataset: config.dataset,
    perspective: config.perspective,
    hasStudioConfig: !!config.studio,
    hasStudioTokenSource: !!config.studio?.auth?.token,
    legacyStudioMode: config.studioMode?.enabled,
    hasAuthProviders: !!config.auth?.providers,
    hasAuthToken: !!config.auth?.token
  });
  const instance = {
    instanceId,
    config,
    isDisposed: () => disposed.current,
    dispose: () => {
      if (disposed.current) {
        logger.trace("Dispose called on already disposed instance", { internal: !0 });
        return;
      }
      logger.trace("Disposing instance", {
        internal: !0,
        listenerCount: disposeListeners.size
      }), disposed.current = !0, disposeListeners.forEach((listener) => listener()), disposeListeners.clear(), logger.info("Instance disposed");
    },
    onDispose: (cb) => {
      const listenerId = insecureRandomId();
      return disposeListeners.set(listenerId, cb), () => {
        disposeListeners.delete(listenerId);
      };
    },
    getParent: () => {
    },
    createChild: (next) => {
      logger.debug("Creating child instance", {
        parentInstanceId: instanceId.slice(0, 8),
        overridingProjectId: !!next.projectId,
        overridingDataset: !!next.dataset,
        overridingAuth: !!next.auth
      });
      const child = Object.assign(
        createSanityInstance({
          ...config,
          ...next,
          ...config.auth === next.auth ? config.auth : config.auth && next.auth && { auth: { ...config.auth, ...next.auth } }
        }),
        { getParent: () => instance }
      );
      return logger.trace("Child instance created", {
        internal: !0,
        childInstanceId: child.instanceId.slice(0, 8)
      }), child;
    },
    match: (targetConfig) => {
      if (Object.entries(pick(targetConfig, "auth", "projectId", "dataset")).every(
        ([key, value]) => config[key] === value
      ))
        return instance;
      const parent = instance.getParent();
      if (parent) return parent.match(targetConfig);
    }
  };
  return instance;
}
function getEnv(key) {
  if (typeof import.meta < "u" && import.meta.env)
    return import.meta.env[key];
  if (typeof process < "u" && process.env)
    return process.env[key];
  if (typeof window < "u" && window.ENV)
    return window.ENV?.[key];
}
function createStoreState(initialState, devToolsOptions) {
  const store = createStore()(devtools(() => initialState, devToolsOptions));
  return {
    get: store.getState,
    set: (actionKey, updatedState) => {
      const currentState = store.getState(), nextState = typeof updatedState == "function" ? updatedState(currentState) : updatedState;
      currentState !== nextState && store.setState(nextState, !1, actionKey);
    },
    observable: new Observable((observer) => {
      const emit = () => observer.next(store.getState());
      emit();
      const unsubscribe = store.subscribe(emit);
      return () => unsubscribe();
    })
  };
}
function createStoreInstance(instance, key, { name, getInitialState: getInitialState2, initialize }) {
  const state = createStoreState(getInitialState2(instance, key), {
    enabled: !!getEnv("DEV"),
    name: `${name}-${key.name}`
  }), dispose = initialize?.({ state, instance, key }), disposed = { current: !1 };
  return {
    state,
    dispose: () => {
      disposed.current || (disposed.current = !0, dispose?.());
    },
    isDisposed: () => disposed.current
  };
}
function createActionBinder(keyFn) {
  const instanceRegistry = /* @__PURE__ */ new Map(), storeRegistry = /* @__PURE__ */ new Map();
  return function(storeDefinition, action) {
    return function(instance, ...params) {
      const key = keyFn(instance, ...params), compositeKey = storeDefinition.name + (key.name ? `:${key.name}` : "");
      let instances = instanceRegistry.get(compositeKey);
      instances || (instances = /* @__PURE__ */ new Set(), instanceRegistry.set(compositeKey, instances)), instances.has(instance.instanceId) || (instances.add(instance.instanceId), instance.onDispose(() => {
        instances.delete(instance.instanceId), instances.size === 0 && (storeRegistry.get(compositeKey)?.dispose(), storeRegistry.delete(compositeKey), instanceRegistry.delete(compositeKey));
      }));
      let storeInstance = storeRegistry.get(compositeKey);
      return storeInstance || (storeInstance = createStoreInstance(instance, key, storeDefinition), storeRegistry.set(compositeKey, storeInstance)), action({ instance, state: storeInstance.state, key }, ...params);
    };
  };
}
const bindActionByDataset = createActionBinder((instance, options) => {
  const projectId = options?.projectId ?? instance.config.projectId, dataset = options?.dataset ?? instance.config.dataset;
  if (!projectId || !dataset)
    throw new Error("This API requires a project ID and dataset configured.");
  return { name: `${projectId}.${dataset}`, projectId, dataset };
}), createSourceKey = (instance, source) => {
  let name, sourceForKey;
  if (source) {
    if (sourceForKey = source, isDatasetSource(source))
      name = `${source.projectId}.${source.dataset}`;
    else if (isMediaLibrarySource(source))
      name = `media-library:${source.mediaLibraryId}`;
    else if (isCanvasSource(source))
      name = `canvas:${source.canvasId}`;
    else
      throw new Error(`Received invalid source: ${JSON.stringify(source)}`);
    return { name, source: sourceForKey };
  }
  const { projectId, dataset } = instance.config;
  if (!projectId || !dataset)
    throw new Error("This API requires a project ID and dataset configured.");
  return { name: `${projectId}.${dataset}`, source: { projectId, dataset } };
}, bindActionBySource = createActionBinder((instance, { source }) => createSourceKey(instance, source)), bindActionBySourceAndPerspective = createActionBinder((instance, options) => {
  const { source, perspective } = options, utilizedPerspective = perspective ?? instance.config.perspective ?? "drafts";
  let perspectiveKey;
  isReleasePerspective(utilizedPerspective) ? perspectiveKey = utilizedPerspective.releaseName : typeof utilizedPerspective == "string" ? perspectiveKey = utilizedPerspective : perspectiveKey = JSON.stringify(utilizedPerspective);
  const sourceKey = createSourceKey(instance, source);
  return {
    name: `${sourceKey.name}:${perspectiveKey}`,
    source: sourceKey.source,
    perspective: utilizedPerspective
  };
}), bindActionGlobally = createActionBinder((..._rest) => ({ name: "global" }));
function createStateSourceAction(options) {
  const selector = typeof options == "function" ? options : options.selector, subscribeHandler = options && "onSubscribe" in options ? options.onSubscribe : void 0, isEqual2 = options && "isEqual" in options ? options.isEqual ?? Object.is : Object.is, selectorContextCache = /* @__PURE__ */ new WeakMap();
  function stateSourceAction(context, ...params) {
    const { state, instance } = context, getCurrent = () => {
      const currentState = state.get();
      if (typeof currentState != "object" || currentState === null)
        throw new Error(
          `Expected store state to be an object but got "${typeof currentState}" instead`
        );
      let instanceCache = selectorContextCache.get(currentState);
      instanceCache || (instanceCache = /* @__PURE__ */ new WeakMap(), selectorContextCache.set(currentState, instanceCache));
      let selectorContext = instanceCache.get(instance);
      return selectorContext || (selectorContext = { state: currentState, instance }, instanceCache.set(instance, selectorContext)), selector(selectorContext, ...params);
    }, subscribe = (onStoreChanged) => {
      const cleanup = subscribeHandler?.(context, ...params), subscription = state.observable.pipe(
        // Derive value from current state
        map(getCurrent),
        // Filter unchanged values using custom equality check
        distinctUntilChanged(isEqual2),
        // Skip initial emission since we only want changes
        skip(1)
      ).subscribe({
        next: () => onStoreChanged?.(),
        // Propagate selector errors to both subscription types
        error: () => onStoreChanged?.()
      });
      return () => {
        subscription.unsubscribe(), cleanup?.();
      };
    }, observable = new Observable((observer) => {
      const emitCurrent = () => {
        try {
          observer.next(getCurrent());
        } catch (error) {
          observer.error(error);
        }
      };
      return emitCurrent(), subscribe(emitCurrent);
    }).pipe(share());
    return {
      getCurrent,
      subscribe,
      observable
    };
  }
  return stateSourceAction;
}
const DEFAULT_BASE = "http://localhost", AUTH_CODE_PARAM = "sid", DEFAULT_API_VERSION$1 = "2021-06-07", REQUEST_TAG_PREFIX = "sanity.sdk.auth";
function resolveAuthMode(config, locationHref) {
  return isStudioConfig(config) ? "studio" : detectDashboardContext(locationHref) ? "dashboard" : "standalone";
}
function isStudioConfig(config) {
  return !!config.studio || !!config.studioMode?.enabled;
}
function detectDashboardContext(locationHref) {
  try {
    const contextParam = new URL(locationHref, DEFAULT_BASE).searchParams.get("_context");
    if (!contextParam) return !1;
    const parsed = JSON.parse(contextParam);
    return typeof parsed == "object" && parsed !== null && !Array.isArray(parsed) && Object.keys(parsed).length > 0;
  } catch (err) {
    return console.error("Failed to parse dashboard context from initial location:", err), !1;
  }
}
var AuthStateType = /* @__PURE__ */ ((AuthStateType2) => (AuthStateType2.LOGGED_IN = "logged-in", AuthStateType2.LOGGING_IN = "logging-in", AuthStateType2.ERROR = "error", AuthStateType2.LOGGED_OUT = "logged-out", AuthStateType2))(AuthStateType || {});
const REFRESH_INTERVAL = 720 * 60 * 1e3, LOCK_NAME = "sanity-token-refresh-lock";
function getLastRefreshTime(storageArea, storageKey) {
  try {
    const data = storageArea?.getItem(`${storageKey}_last_refresh`), parsed = data ? parseInt(data, 10) : 0;
    return isNaN(parsed) ? 0 : parsed;
  } catch {
    return 0;
  }
}
function setLastRefreshTime(storageArea, storageKey) {
  try {
    storageArea?.setItem(`${storageKey}_last_refresh`, Date.now().toString());
  } catch {
  }
}
function getNextRefreshDelay(storageArea, storageKey) {
  const lastRefresh = getLastRefreshTime(storageArea, storageKey);
  if (!lastRefresh) return 0;
  const now = Date.now(), nextRefreshTime = lastRefresh + REFRESH_INTERVAL;
  return Math.max(0, nextRefreshTime - now);
}
function createTokenRefreshStream(token, clientFactory, apiHost) {
  return new Observable((subscriber) => {
    const subscription = clientFactory({
      apiVersion: DEFAULT_API_VERSION$1,
      requestTagPrefix: "token-refresh",
      useProjectHostname: !1,
      useCdn: !1,
      token,
      ignoreBrowserTokenWarning: !0,
      ...apiHost && { apiHost }
    }).observable.request({
      uri: "auth/refresh-token",
      method: "POST",
      body: {
        token
      }
    }).subscribe(subscriber);
    return () => subscription.unsubscribe();
  });
}
async function acquireTokenRefreshLock(refreshFn, storageArea, storageKey) {
  if (!navigator.locks)
    return console.warn("Web Locks API not supported. Proceeding with uncoordinated refresh."), await refreshFn(), setLastRefreshTime(storageArea, storageKey), !0;
  try {
    return await navigator.locks.request(LOCK_NAME, { mode: "exclusive" }, async (lock) => {
      if (!lock) return !1;
      for (; ; ) {
        const delay2 = getNextRefreshDelay(storageArea, storageKey);
        delay2 > 0 && await new Promise((resolve) => setTimeout(resolve, delay2));
        try {
          await refreshFn(), setLastRefreshTime(storageArea, storageKey);
        } catch (error) {
          console.error("Token refresh failed within lock:", error);
        }
        await new Promise((resolve) => setTimeout(resolve, REFRESH_INTERVAL));
      }
    }) === !0;
  } catch (error) {
    return console.error("Failed to request token refresh lock:", error), !1;
  }
}
function shouldRefreshToken(lastRefresh) {
  return lastRefresh ? Date.now() - lastRefresh >= REFRESH_INTERVAL : !0;
}
const refreshStampedToken = ({ state }) => {
  const { clientFactory, apiHost, storageArea, storageKey } = state.get().options;
  return state.observable.pipe(
    map((storeState) => ({
      authState: storeState.authState,
      dashboardContext: storeState.dashboardContext
    })),
    filter(
      (storeState) => storeState.authState.type === AuthStateType.LOGGED_IN
    ),
    distinctUntilChanged(
      (prev, curr) => prev.authState.type === curr.authState.type && prev.authState.token === curr.authState.token && // Only care about token for distinctness here
      prev.dashboardContext === curr.dashboardContext
    ),
    // Make distinctness check explicit
    filter((storeState) => storeState.authState.token.includes("-st")),
    // Ensure we only try to refresh stamped tokens
    exhaustMap((storeState) => {
      const performRefresh = async () => {
        const currentState = state.get();
        if (currentState.authState.type !== AuthStateType.LOGGED_IN)
          throw new Error("User logged out before refresh could complete");
        const currentToken = currentState.authState.token, response = await firstValueFrom(
          createTokenRefreshStream(currentToken, clientFactory, apiHost)
        );
        state.set("setRefreshStampedToken", (prev) => ({
          authState: prev.authState.type === AuthStateType.LOGGED_IN ? { ...prev.authState, token: response.token } : prev.authState
        })), storageArea?.setItem(storageKey, JSON.stringify({ token: response.token }));
      };
      return storeState.dashboardContext ? new Observable((subscriber) => {
        const visibilityHandler = () => {
          const currentState = state.get();
          document.visibilityState === "visible" && currentState.authState.type === AuthStateType.LOGGED_IN && shouldRefreshToken(currentState.authState.lastTokenRefresh) && createTokenRefreshStream(
            currentState.authState.token,
            clientFactory,
            apiHost
          ).subscribe({
            next: (response) => {
              state.set("setRefreshStampedToken", (prev) => ({
                authState: prev.authState.type === AuthStateType.LOGGED_IN ? {
                  ...prev.authState,
                  token: response.token,
                  lastTokenRefresh: Date.now()
                } : prev.authState
              })), subscriber.next(response);
            },
            error: (error) => subscriber.error(error)
          });
        }, timerSubscription = timer(REFRESH_INTERVAL, REFRESH_INTERVAL).pipe(
          filter(() => document.visibilityState === "visible"),
          switchMap(() => {
            const currentState = state.get().authState;
            if (currentState.type !== AuthStateType.LOGGED_IN)
              throw new Error("User logged out before refresh could complete");
            return createTokenRefreshStream(currentState.token, clientFactory, apiHost);
          })
        ).subscribe({
          next: (response) => {
            state.set("setRefreshStampedToken", (prev) => ({
              authState: prev.authState.type === AuthStateType.LOGGED_IN ? {
                ...prev.authState,
                token: response.token,
                lastTokenRefresh: Date.now()
              } : prev.authState
            })), subscriber.next(response);
          },
          error: (error) => subscriber.error(error)
        });
        return document.addEventListener("visibilitychange", visibilityHandler), () => {
          document.removeEventListener("visibilitychange", visibilityHandler), timerSubscription.unsubscribe();
        };
      }).pipe(
        takeWhile(() => state.get().authState.type === AuthStateType.LOGGED_IN),
        map((response) => ({ token: response.token }))
      ) : from(acquireTokenRefreshLock(performRefresh, storageArea, storageKey)).pipe(
        filter((hasLock) => hasLock),
        map(() => {
          const currentState = state.get().authState;
          if (currentState.type !== AuthStateType.LOGGED_IN)
            throw new Error("User logged out before refresh could complete");
          return { token: currentState.token };
        })
      );
    })
  ).subscribe({
    next: (response) => {
      state.set("setRefreshStampedToken", (prev) => ({
        authState: prev.authState.type === AuthStateType.LOGGED_IN ? {
          ...prev.authState,
          token: response.token,
          lastTokenRefresh: Date.now()
        } : prev.authState
      })), storageArea?.setItem(storageKey, JSON.stringify({ token: response.token }));
    },
    error: (error) => {
      state.set("setRefreshStampedTokenError", { authState: { type: AuthStateType.ERROR, error } });
    }
  });
}, subscribeToStateAndFetchCurrentUser = ({ state, instance }, fetchOptions) => {
  const { clientFactory, apiHost } = state.get().options, useProjectHostname = fetchOptions?.useProjectHostname ?? isStudioConfig(instance.config), projectId = instance.config.projectId;
  return state.observable.pipe(
    map(({ authState, options: storeOptions }) => ({
      authState,
      authMethod: storeOptions.authMethod
    })),
    filter(
      (value) => value.authState.type === AuthStateType.LOGGED_IN && !value.authState.currentUser
    ),
    map((value) => ({ token: value.authState.token, authMethod: value.authMethod })),
    distinctUntilChanged(
      (prev, curr) => prev.token === curr.token && prev.authMethod === curr.authMethod
    )
  ).pipe(
    map(
      ({ token, authMethod }) => clientFactory({
        apiVersion: DEFAULT_API_VERSION$1,
        requestTagPrefix: REQUEST_TAG_PREFIX,
        token: authMethod === "cookie" ? void 0 : token,
        ignoreBrowserTokenWarning: !0,
        useProjectHostname,
        useCdn: !1,
        ...authMethod === "cookie" ? { withCredentials: !0 } : {},
        ...useProjectHostname && projectId ? { projectId } : {},
        ...apiHost && { apiHost }
      })
    ),
    switchMap(
      (client) => client.observable.request({ uri: "/users/me", method: "GET" })
    )
  ).subscribe({
    next: (currentUser) => {
      state.set("setCurrentUser", (prev) => ({
        authState: prev.authState.type === AuthStateType.LOGGED_IN ? { ...prev.authState, currentUser } : prev.authState
      }));
    },
    error: (error) => {
      state.set("setError", { authState: { type: AuthStateType.ERROR, error } });
    }
  });
};
function getAuthCode(callbackUrl, locationHref) {
  const loc = new URL(locationHref, DEFAULT_BASE), callbackLocation = callbackUrl ? new URL(callbackUrl, DEFAULT_BASE) : void 0, callbackLocationMatches = callbackLocation ? loc.pathname.toLowerCase().startsWith(callbackLocation.pathname.toLowerCase()) : !0;
  let authCode = new URLSearchParams(loc.hash.slice(1)).get(AUTH_CODE_PARAM) || new URLSearchParams(loc.search).get(AUTH_CODE_PARAM);
  if (!authCode) {
    const contextParam = new URLSearchParams(loc.search).get("_context");
    if (contextParam)
      try {
        const parsedContext = JSON.parse(contextParam);
        parsedContext && typeof parsedContext == "object" && typeof parsedContext.sid == "string" && parsedContext.sid && (authCode = parsedContext.sid);
      } catch {
      }
  }
  return authCode && callbackLocationMatches ? authCode : null;
}
function getTokenFromLocation(locationHref) {
  const loc = new URL(locationHref);
  return new URLSearchParams(loc.hash.slice(1)).get("token") || null;
}
function getTokenFromStorage(storageArea, storageKey) {
  if (!storageArea) return null;
  const item = storageArea.getItem(storageKey);
  if (item === null) return null;
  try {
    const parsed = JSON.parse(item);
    if (typeof parsed != "object" || parsed === null || !("token" in parsed) || typeof parsed.token != "string")
      throw new Error("Invalid stored auth data structure");
    return parsed.token;
  } catch {
    return storageArea.removeItem(storageKey), null;
  }
}
function getStorageEvents() {
  return typeof window < "u" && typeof window.addEventListener == "function" ? fromEvent(window, "storage") : EMPTY;
}
function getDefaultStorage() {
  try {
    return typeof localStorage < "u" && typeof localStorage.getItem == "function" ? localStorage : void 0;
  } catch {
    return;
  }
}
function getDefaultLocation() {
  try {
    return typeof location > "u" ? DEFAULT_BASE : typeof location.href == "string" ? location.href : DEFAULT_BASE;
  } catch {
    return DEFAULT_BASE;
  }
}
function getCleanedUrl(locationUrl) {
  const loc = new URL(locationUrl), rawHash = loc.hash.startsWith("#") ? loc.hash.slice(1) : loc.hash;
  if (rawHash && rawHash.includes("=")) {
    const hashParams = new URLSearchParams(rawHash);
    hashParams.delete("token"), hashParams.delete("withSid");
    const nextHash = hashParams.toString();
    loc.hash = nextHash ? `#${nextHash}` : "";
  }
  return loc.searchParams.delete("sid"), loc.searchParams.delete("url"), loc.toString();
}
function getClientErrorApiBody(error) {
  const body = error.response?.body;
  return body && typeof body == "object" ? body : void 0;
}
function getClientErrorApiType(error) {
  const body = getClientErrorApiBody(error);
  return body?.error?.type ?? body?.type;
}
function getClientErrorApiDescription(error) {
  const body = getClientErrorApiBody(error);
  return body?.error?.description ?? body?.description;
}
function isProjectUserNotFoundClientError(error) {
  return getClientErrorApiType(error) === "projectUserNotFoundError";
}
function parseDashboardContext(locationHref) {
  try {
    const contextParam = new URL(locationHref, DEFAULT_BASE).searchParams.get("_context");
    if (contextParam) {
      const parsedContext = JSON.parse(contextParam);
      if (parsedContext && typeof parsedContext == "object" && !Array.isArray(parsedContext) && Object.keys(parsedContext).length > 0)
        return delete parsedContext.sid, parsedContext;
    }
  } catch (err) {
    console.error("Failed to parse dashboard context from initial location:", err);
  }
  return {};
}
function getDashboardInitialState(options) {
  const { authConfig, initialLocationHref } = options, providedToken = authConfig.token, callbackUrl = authConfig.callbackUrl, storageKey = "__sanity_auth_token", dashboardContext = parseDashboardContext(initialLocationHref), storageArea = void 0;
  return providedToken ? {
    authState: { type: AuthStateType.LOGGED_IN, token: providedToken, currentUser: null },
    storageKey,
    storageArea,
    authMethod: void 0,
    dashboardContext
  } : getAuthCode(callbackUrl, initialLocationHref) || getTokenFromLocation(initialLocationHref) ? {
    authState: { type: AuthStateType.LOGGING_IN, isExchangingToken: !1 },
    storageKey,
    storageArea,
    authMethod: void 0,
    dashboardContext
  } : {
    authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 },
    storageKey,
    storageArea,
    authMethod: void 0,
    dashboardContext
  };
}
function initializeDashboardAuth(context, tokenRefresherRunning2) {
  const subscriptions = [];
  let startedRefresher = !1;
  return subscriptions.push(subscribeToStateAndFetchCurrentUser(context, { useProjectHostname: !1 })), tokenRefresherRunning2 || (startedRefresher = !0, subscriptions.push(refreshStampedToken(context))), {
    dispose: () => {
      for (const subscription of subscriptions)
        subscription.unsubscribe();
    },
    tokenRefresherStarted: startedRefresher
  };
}
const subscribeToStorageEventsAndSetToken = ({
  state
}) => {
  const { storageArea, storageKey } = state.get().options;
  return defer(getStorageEvents).pipe(
    filter(
      (e) => e.storageArea === storageArea && e.key === storageKey
    ),
    map(() => getTokenFromStorage(storageArea, storageKey)),
    distinctUntilChanged()
  ).subscribe((token) => {
    state.set("updateTokenFromStorageEvent", {
      authState: token ? { type: AuthStateType.LOGGED_IN, token, currentUser: null } : { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 }
    });
  });
};
function getStandaloneInitialState(options) {
  const { authConfig, initialLocationHref } = options, providedToken = authConfig.token, callbackUrl = authConfig.callbackUrl, storageKey = "__sanity_auth_token", storageArea = authConfig.storageArea ?? getDefaultStorage();
  if (providedToken)
    return {
      authState: { type: AuthStateType.LOGGED_IN, token: providedToken, currentUser: null },
      storageKey,
      storageArea,
      authMethod: void 0,
      dashboardContext: {}
    };
  if (getAuthCode(callbackUrl, initialLocationHref) || getTokenFromLocation(initialLocationHref))
    return {
      authState: { type: AuthStateType.LOGGING_IN, isExchangingToken: !1 },
      storageKey,
      storageArea,
      authMethod: void 0,
      dashboardContext: {}
    };
  const token = getTokenFromStorage(storageArea, storageKey);
  return token ? {
    authState: { type: AuthStateType.LOGGED_IN, token, currentUser: null },
    storageKey,
    storageArea,
    authMethod: "localstorage",
    dashboardContext: {}
  } : {
    authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 },
    storageKey,
    storageArea,
    authMethod: void 0,
    dashboardContext: {}
  };
}
function initializeStandaloneAuth(context, tokenRefresherRunning2) {
  const subscriptions = [];
  let startedRefresher = !1;
  return subscriptions.push(subscribeToStateAndFetchCurrentUser(context, { useProjectHostname: !1 })), context.state.get().options?.storageArea && subscriptions.push(subscribeToStorageEventsAndSetToken(context)), tokenRefresherRunning2 || (startedRefresher = !0, subscriptions.push(refreshStampedToken(context))), {
    dispose: () => {
      for (const subscription of subscriptions)
        subscription.unsubscribe();
    },
    tokenRefresherStarted: startedRefresher
  };
}
const COOKIE_AUTH_TIMEOUT_MS = 1e4;
async function checkForCookieAuth(projectId, clientFactory) {
  if (!projectId) return !1;
  try {
    const user = await clientFactory({
      projectId,
      useCdn: !1,
      requestTagPrefix: "sdk",
      timeout: COOKIE_AUTH_TIMEOUT_MS
    }).request({
      uri: "/users/me",
      withCredentials: !0,
      tag: "users.get-current"
    });
    return user != null && typeof user == "object" && typeof user.id == "string";
  } catch {
    return !1;
  }
}
function getStudioTokenFromLocalStorage(storageArea, storageKey) {
  return !storageArea || !storageKey ? null : getTokenFromStorage(storageArea, storageKey) || null;
}
function getStudioInitialState(options) {
  const { authConfig, projectId, tokenSource } = options, storageArea = authConfig.storageArea ?? getDefaultStorage(), studioStorageKey = `__studio_auth_token_${projectId ?? ""}`;
  if (tokenSource)
    return {
      authState: { type: AuthStateType.LOGGING_IN, isExchangingToken: !1 },
      storageKey: studioStorageKey,
      storageArea,
      authMethod: void 0,
      dashboardContext: {}
    };
  const providedToken = authConfig.token;
  let authMethod;
  const token = getStudioTokenFromLocalStorage(storageArea, studioStorageKey);
  return token && (authMethod = "localstorage"), providedToken ? {
    authState: { type: AuthStateType.LOGGED_IN, token: providedToken, currentUser: null },
    storageKey: studioStorageKey,
    storageArea,
    authMethod,
    dashboardContext: {}
  } : token ? {
    authState: { type: AuthStateType.LOGGED_IN, token, currentUser: null },
    storageKey: studioStorageKey,
    storageArea,
    authMethod: "localstorage",
    dashboardContext: {}
  } : {
    authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 },
    storageKey: studioStorageKey,
    storageArea,
    authMethod: void 0,
    dashboardContext: {}
  };
}
function initializeStudioAuth(context, tokenRefresherRunning2) {
  const tokenSource = context.instance.config.studio?.auth?.token;
  return tokenSource ? initializeWithTokenSource(context, tokenSource) : initializeWithFallback(context, tokenRefresherRunning2);
}
function initializeWithTokenSource(context, tokenSource) {
  const subscriptions = [];
  subscriptions.push(subscribeToStateAndFetchCurrentUser(context, { useProjectHostname: !0 }));
  const tokenSub = tokenSource.subscribe({
    next: (token) => {
      const { state } = context;
      token ? state.set("studioTokenSource", (prev) => ({
        options: { ...prev.options, authMethod: void 0 },
        authState: { type: AuthStateType.LOGGED_IN, token, currentUser: null }
      })) : state.set("studioTokenSourceLoggedOut", (prev) => ({
        options: { ...prev.options, authMethod: void 0 },
        authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 }
      }));
    }
  });
  return {
    dispose: () => {
      tokenSub.unsubscribe();
      for (const subscription of subscriptions)
        subscription.unsubscribe();
    },
    // Studio handles token refresh — do not start the SDK's refresher
    tokenRefresherStarted: !1
  };
}
function initializeWithFallback(context, tokenRefresherRunning2) {
  const subscriptions = [];
  let startedRefresher = !1;
  subscriptions.push(subscribeToStateAndFetchCurrentUser(context, { useProjectHostname: !0 })), context.state.get().options?.storageArea && subscriptions.push(subscribeToStorageEventsAndSetToken(context));
  try {
    const { instance, state } = context;
    if (!(state.get().authState?.type === AuthStateType.LOGGED_IN && state.get().authState.token)) {
      const projectIdValue = instance.config.projectId, clientFactory = state.get().options.clientFactory;
      checkForCookieAuth(projectIdValue, clientFactory).then((isCookieAuthEnabled) => {
        isCookieAuthEnabled && state.set("enableCookieAuth", (prev) => ({
          options: { ...prev.options, authMethod: "cookie" },
          authState: prev.authState.type === AuthStateType.LOGGED_IN ? prev.authState : { type: AuthStateType.LOGGED_IN, token: "", currentUser: null }
        }));
      });
    }
  } catch {
  }
  return tokenRefresherRunning2 || (startedRefresher = !0, subscriptions.push(refreshStampedToken(context))), {
    dispose: () => {
      for (const subscription of subscriptions)
        subscription.unsubscribe();
    },
    tokenRefresherStarted: startedRefresher
  };
}
let tokenRefresherRunning = !1;
const authStore = {
  name: "Auth",
  getInitialState(instance) {
    const {
      apiHost,
      callbackUrl,
      providers: customProviders,
      token: providedToken,
      clientFactory = createClient,
      initialLocationHref = getDefaultLocation()
    } = instance.config.auth ?? {}, authConfig = instance.config.auth ?? {};
    let loginDomain = "https://www.sanity.io";
    try {
      apiHost && new URL(apiHost).hostname.endsWith(".sanity.work") && (loginDomain = "https://www.sanity.work");
    } catch {
    }
    const loginUrl = new URL("/login", loginDomain);
    loginUrl.searchParams.set("origin", getCleanedUrl(initialLocationHref)), loginUrl.searchParams.set("type", "stampedToken"), loginUrl.searchParams.set("withSid", "true");
    const mode = resolveAuthMode(instance.config, initialLocationHref), strategyOptions = {
      authConfig,
      projectId: instance.config.projectId,
      initialLocationHref,
      tokenSource: instance.config.studio?.auth?.token
    };
    let result;
    switch (mode) {
      case "studio":
        result = getStudioInitialState(strategyOptions);
        break;
      case "dashboard":
        result = getDashboardInitialState(strategyOptions);
        break;
      case "standalone":
        result = getStandaloneInitialState(strategyOptions);
        break;
    }
    return {
      authState: result.authState,
      dashboardContext: result.dashboardContext,
      options: {
        apiHost,
        loginUrl: loginUrl.toString(),
        callbackUrl,
        customProviders,
        providedToken,
        clientFactory,
        initialLocationHref,
        storageKey: result.storageKey,
        storageArea: result.storageArea,
        authMethod: result.authMethod
      }
    };
  },
  initialize(context) {
    const initialLocationHref = context.state.get().options?.initialLocationHref ?? getDefaultLocation(), mode = resolveAuthMode(context.instance.config, initialLocationHref);
    let initResult;
    switch (mode) {
      case "studio":
        initResult = initializeStudioAuth(context, tokenRefresherRunning);
        break;
      case "dashboard":
        initResult = initializeDashboardAuth(context, tokenRefresherRunning);
        break;
      case "standalone":
        initResult = initializeStandaloneAuth(context, tokenRefresherRunning);
        break;
    }
    return initResult.tokenRefresherStarted && (tokenRefresherRunning = !0), initResult.dispose;
  }
}, getCurrentUserState = bindActionGlobally(
  authStore,
  createStateSourceAction(
    ({ state: { authState } }) => authState.type === AuthStateType.LOGGED_IN ? authState.currentUser : null
  )
), getTokenState = bindActionGlobally(
  authStore,
  createStateSourceAction(
    ({ state: { authState } }) => authState.type === AuthStateType.LOGGED_IN ? authState.token : null
  )
), getAuthMethodState = bindActionGlobally(
  authStore,
  createStateSourceAction(({ state: { options } }) => options.authMethod)
), getLoginUrlState = bindActionGlobally(
  authStore,
  createStateSourceAction(({ state: { options } }) => options.loginUrl)
), getAuthState = bindActionGlobally(
  authStore,
  createStateSourceAction(({ state: { authState } }) => authState)
), getDashboardOrganizationId$1 = bindActionGlobally(
  authStore,
  createStateSourceAction(({ state: { dashboardContext } }) => dashboardContext?.orgId)
), getIsInDashboardState = bindActionGlobally(
  authStore,
  createStateSourceAction(
    ({ state: { dashboardContext } }) => (
      // Check if dashboardContext exists and is not empty
      !!dashboardContext && Object.keys(dashboardContext).length > 0
    )
  )
), setAuthToken = bindActionGlobally(authStore, ({ state }, token) => {
  const currentAuthState = state.get().authState;
  token ? (currentAuthState.type !== AuthStateType.LOGGED_IN || currentAuthState.token !== token) && state.set("setToken", {
    authState: {
      type: AuthStateType.LOGGED_IN,
      token,
      // Keep existing user or set to null? Setting to null forces refetch.
      // Keep existing user to avoid unnecessary refetches if user data is still valid.
      currentUser: currentAuthState.type === AuthStateType.LOGGED_IN ? currentAuthState.currentUser : null
    }
  }) : currentAuthState.type !== AuthStateType.LOGGED_OUT && state.set("setToken", {
    authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 }
  });
}), DEFAULT_API_VERSION = "2024-11-12", DEFAULT_REQUEST_TAG_PREFIX = "sanity.sdk", allowedKeys = Object.keys({
  apiHost: null,
  useCdn: null,
  token: null,
  perspective: null,
  proxy: null,
  withCredentials: null,
  timeout: null,
  maxRetries: null,
  dataset: null,
  projectId: null,
  scope: null,
  apiVersion: null,
  requestTagPrefix: null,
  useProjectHostname: null,
  "~experimental_resource": null,
  source: null
}), DEFAULT_CLIENT_CONFIG = {
  apiVersion: DEFAULT_API_VERSION,
  useCdn: !1,
  ignoreBrowserTokenWarning: !0,
  allowReconfigure: !1,
  requestTagPrefix: DEFAULT_REQUEST_TAG_PREFIX
}, clientStore = {
  name: "clientStore",
  getInitialState: (instance) => ({
    clients: {},
    token: getTokenState(instance).getCurrent()
  }),
  initialize(context) {
    const subscription = listenToToken(context), authMethodSubscription = listenToAuthMethod(context);
    return () => {
      subscription.unsubscribe(), authMethodSubscription.unsubscribe();
    };
  }
}, listenToToken = ({ instance, state }) => getTokenState(instance).observable.subscribe((token) => {
  state.set("setTokenAndResetClients", { token, clients: {} });
}), listenToAuthMethod = ({ instance, state }) => getAuthMethodState(instance).observable.subscribe((authMethod) => {
  state.set("setAuthMethod", { authMethod });
}), getClientConfigKey = (options) => JSON.stringify(pick(options, ...allowedKeys)), getClient = bindActionGlobally(
  clientStore,
  ({ state, instance }, options) => {
    if (!options || typeof options != "object")
      throw new Error(
        'getClient() requires a configuration object with at least an "apiVersion" property. Example: getClient(instance, { apiVersion: "2024-11-12" })'
      );
    const disallowedKeys = Object.keys(options).filter((key2) => !allowedKeys.includes(key2));
    if (disallowedKeys.length > 0) {
      const listFormatter = new Intl.ListFormat("en", { style: "long", type: "conjunction" });
      throw new Error(
        `The client options provided contains unsupported properties: ${listFormatter.format(disallowedKeys)}. Allowed keys are: ${listFormatter.format(allowedKeys)}.`
      );
    }
    const tokenFromState = state.get().token, { clients, authMethod } = state.get();
    let resource;
    options.source && (isDatasetSource(options.source) ? resource = { type: "dataset", id: `${options.source.projectId}.${options.source.dataset}` } : isMediaLibrarySource(options.source) ? resource = { type: "media-library", id: options.source.mediaLibraryId } : isCanvasSource(options.source) && (resource = { type: "canvas", id: options.source.canvasId }));
    const projectId = options.projectId ?? instance.config.projectId, dataset = options.dataset ?? instance.config.dataset, apiHost = options.apiHost ?? instance.config.auth?.apiHost, effectiveOptions = {
      ...DEFAULT_CLIENT_CONFIG,
      ...(options.scope === "global" || !projectId || resource) && { useProjectHostname: !1 },
      token: authMethod === "cookie" ? void 0 : tokenFromState ?? void 0,
      ...options,
      ...projectId && { projectId },
      ...dataset && { dataset },
      ...apiHost && { apiHost },
      ...resource && { "~experimental_resource": resource }
    };
    resource && ((options.projectId || options.dataset) && console.warn(
      "Both source and explicit projectId/dataset are provided. The source will be used and projectId/dataset will be ignored."
    ), delete effectiveOptions.projectId, delete effectiveOptions.dataset), effectiveOptions.token === null || typeof effectiveOptions.token > "u" ? (delete effectiveOptions.token, authMethod === "cookie" && (effectiveOptions.withCredentials = !0)) : delete effectiveOptions.withCredentials;
    const key = getClientConfigKey(effectiveOptions);
    if (clients[key]) return clients[key];
    const client = createClient(effectiveOptions);
    return state.set("addClient", (prev) => ({ clients: { ...prev.clients, [key]: client } })), client;
  }
), getClientState = bindActionGlobally(
  clientStore,
  createStateSourceAction(({ instance }, options) => getClient(instance, options))
), API_VERSION$6 = "vX";
function agentGenerate(instance, options) {
  return getClientState(instance, {
    apiVersion: API_VERSION$6,
    projectId: instance.config.projectId,
    dataset: instance.config.dataset
  }).observable.pipe(switchMap((client) => client.observable.agent.action.generate(options)));
}
function agentTransform(instance, options) {
  return getClientState(instance, {
    apiVersion: API_VERSION$6,
    projectId: instance.config.projectId,
    dataset: instance.config.dataset
  }).observable.pipe(switchMap((client) => client.observable.agent.action.transform(options)));
}
function agentTranslate(instance, options) {
  return getClientState(instance, {
    apiVersion: API_VERSION$6,
    projectId: instance.config.projectId,
    dataset: instance.config.dataset
  }).observable.pipe(switchMap((client) => client.observable.agent.action.translate(options)));
}
function agentPrompt(instance, options) {
  return getClientState(instance, {
    apiVersion: API_VERSION$6,
    projectId: instance.config.projectId,
    dataset: instance.config.dataset
  }).observable.pipe(switchMap((client) => from(client.agent.action.prompt(options))));
}
function agentPatch(instance, options) {
  return getClientState(instance, {
    apiVersion: API_VERSION$6,
    projectId: instance.config.projectId,
    dataset: instance.config.dataset
  }).observable.pipe(switchMap((client) => from(client.agent.action.patch(options))));
}
function compareProjectOrganization(projectId, projectOrganizationId, currentDashboardOrgId) {
  return projectOrganizationId !== currentDashboardOrgId ? {
    error: `Project ${projectId} belongs to Organization ${projectOrganizationId ?? "unknown"}, but the Dashboard has Organization ${currentDashboardOrgId} selected`
  } : { error: null };
}
function createFetcherStore({
  name,
  fetcher: getObservable,
  getKey,
  fetchThrottleInternal = 1e3,
  stateExpirationDelay = 5e3
}) {
  const store = {
    name,
    getInitialState: () => ({
      stateByParams: {}
    }),
    initialize: (context) => {
      const subscription = subscribeToSubscriptionsAndFetch(context);
      return () => subscription.unsubscribe();
    }
  }, subscribeToSubscriptionsAndFetch = ({
    state
  }) => state.observable.pipe(
    // Map the state to an array of [serialized, entry] pairs.
    switchMap$1((s) => {
      const entries = Object.entries(s.stateByParams);
      return entries.length > 0 ? from(entries) : EMPTY;
    }),
    // Group by the serialized key.
    groupBy(([key]) => key),
    mergeMap(
      (group$) => group$.pipe(
        // Emit an initial value for pairwise comparisons.
        startWith([group$.key, void 0]),
        pairwise(),
        // Trigger only when the subscriptions array grows.
        filter$1(([[, prevEntry], [, currEntry]]) => {
          const prevSubs = prevEntry?.subscriptions ?? [];
          return (currEntry?.subscriptions ?? []).length > prevSubs.length;
        }),
        map$1(([, [, currEntry]]) => currEntry),
        // Only trigger if we haven't fetched recently.
        filter$1((entry) => {
          const lastFetch = entry?.lastFetchInitiatedAt;
          return lastFetch ? Date.now() - new Date(lastFetch).getTime() >= fetchThrottleInternal : !0;
        }),
        switchMap$1((entry) => entry ? (state.set("setLastFetchInitiatedAt", (prev) => ({
          stateByParams: {
            ...prev.stateByParams,
            [entry.key]: {
              ...entry,
              ...prev.stateByParams[entry.key],
              lastFetchInitiatedAt: (/* @__PURE__ */ new Date()).toISOString()
            }
          }
        })), getObservable(entry.instance)(...entry.params).pipe(
          // the `createStateSourceAction` util requires the update
          // to
          delay(0, asapScheduler),
          tap(
            (data) => state.set("setData", (prev) => ({
              stateByParams: {
                ...prev.stateByParams,
                [entry.key]: {
                  ...omit(entry, "error"),
                  ...omit(prev.stateByParams[entry.key], "error"),
                  data
                }
              }
            }))
          ),
          catchError((error) => (state.set("setError", (prev) => ({
            stateByParams: {
              ...prev.stateByParams,
              [entry.key]: {
                ...entry,
                ...prev.stateByParams[entry.key],
                error
              }
            }
          })), EMPTY))
        )) : EMPTY)
      )
    )
  ).subscribe({
    error: (error) => state.set("setError", { error })
  }), getState = bindActionGlobally(
    store,
    createStateSourceAction({
      selector: ({
        instance,
        state: { stateByParams, error }
      }, ...params) => {
        if (error) throw error;
        const key = getKey(instance, ...params), entry = stateByParams[key];
        if (entry?.error) throw entry.error;
        return entry?.data;
      },
      onSubscribe: ({ instance, state }, ...params) => {
        const subscriptionId = insecureRandomId(), key = getKey(instance, ...params);
        return state.set("addSubscription", (prev) => ({
          stateByParams: {
            ...prev.stateByParams,
            [key]: {
              ...prev.stateByParams[key],
              instance,
              key,
              params: prev.stateByParams[key]?.params || params,
              subscriptions: [...prev.stateByParams[key]?.subscriptions || [], subscriptionId]
            }
          }
        })), () => {
          setTimeout(() => {
            state.set("removeSubscription", (prev) => {
              const entry = prev.stateByParams[key];
              if (!entry) return prev;
              const newSubs = (entry.subscriptions || []).filter((id) => id !== subscriptionId);
              return newSubs.length === 0 ? { stateByParams: omit(prev.stateByParams, key) } : {
                stateByParams: {
                  ...prev.stateByParams,
                  [key]: {
                    ...entry,
                    subscriptions: newSubs
                  }
                }
              };
            });
          }, stateExpirationDelay);
        };
      }
    })
  ), resolveState = bindActionGlobally(
    store,
    ({ instance }, ...params) => firstValueFrom(getState(instance, ...params).observable.pipe(first((i) => i !== void 0)))
  );
  return { getState, resolveState };
}
const API_VERSION$5 = "v2025-02-19", project = createFetcherStore({
  name: "Project",
  getKey: (instance, options) => {
    const projectId = options?.projectId ?? instance.config.projectId;
    if (!projectId)
      throw new Error("A projectId is required to use the project API.");
    return projectId;
  },
  fetcher: (instance) => (options = {}) => {
    const projectId = options.projectId ?? instance.config.projectId;
    return getClientState(instance, {
      apiVersion: API_VERSION$5,
      scope: "global",
      projectId
    }).observable.pipe(
      switchMap(
        (client) => client.observable.projects.getById(
          // non-null assertion is fine with the above throwing
          projectId ?? instance.config.projectId
        )
      )
    );
  }
}), getProjectState = project.getState, resolveProject = project.resolveState, getDashboardOrganizationId = bindActionGlobally(
  authStore,
  createStateSourceAction(({ state: { dashboardContext } }) => dashboardContext?.orgId)
);
function observeOrganizationVerificationState(instance, projectIds) {
  const dashboardOrgId$ = getDashboardOrganizationId(instance).observable.pipe(distinctUntilChanged()), projectOrgIdObservables = projectIds.map(
    (id) => getProjectState(instance, { projectId: id }).observable.pipe(
      map((project2) => ({ projectId: id, orgId: project2?.organizationId ?? null })),
      // Ensure we only proceed if the orgId is loaded, distinct prevents unnecessary checks
      distinctUntilChanged((prev, curr) => prev.orgId === curr.orgId)
    )
  ), allProjectOrgIds$ = projectOrgIdObservables.length > 0 ? combineLatest(projectOrgIdObservables) : of([]);
  return combineLatest([dashboardOrgId$, allProjectOrgIds$]).pipe(
    switchMap(([dashboardOrgId, projectOrgDataArray]) => {
      if (!dashboardOrgId || projectOrgDataArray.length === 0)
        return of({ error: null });
      for (const projectData of projectOrgDataArray) {
        if (!projectData.orgId)
          continue;
        const result = compareProjectOrganization(
          projectData.projectId,
          projectData.orgId,
          dashboardOrgId
        );
        if (result.error)
          return of(result);
      }
      return of({ error: null });
    }),
    // Only emit when the overall error status actually changes
    distinctUntilChanged((prev, curr) => prev.error === curr.error)
  );
}
const handleAuthCallback = bindActionGlobally(
  authStore,
  async ({ state }, locationHref = getDefaultLocation()) => {
    const { providedToken, callbackUrl, clientFactory, apiHost, storageArea, storageKey } = state.get().options;
    if (providedToken) return !1;
    const { authState } = state.get();
    if (authState.type === AuthStateType.LOGGING_IN && authState.isExchangingToken) return !1;
    const cleanedUrl = getCleanedUrl(locationHref), tokenFromUrl = getTokenFromLocation(locationHref);
    if (tokenFromUrl)
      return state.set("setTokenFromUrl", {
        authState: { type: AuthStateType.LOGGED_IN, token: tokenFromUrl, currentUser: null }
      }), cleanedUrl;
    const authCode = getAuthCode(callbackUrl, locationHref);
    if (!authCode) return !1;
    const parsedUrl = new URL(locationHref);
    let dashboardContext = {};
    try {
      const contextParam = parsedUrl.searchParams.get("_context");
      if (contextParam) {
        const parsedContext = JSON.parse(contextParam);
        parsedContext && typeof parsedContext == "object" && (delete parsedContext.sid, dashboardContext = parsedContext);
      }
    } catch (err) {
      console.error("Failed to parse dashboard context:", err);
    }
    state.set("exchangeSessionForToken", {
      authState: { type: AuthStateType.LOGGING_IN, isExchangingToken: !0 },
      dashboardContext
    });
    try {
      const client = clientFactory({
        apiVersion: DEFAULT_API_VERSION$1,
        requestTagPrefix: REQUEST_TAG_PREFIX,
        useProjectHostname: !1,
        useCdn: !1,
        ...apiHost && { apiHost }
      }), { token } = await client.request({
        method: "GET",
        uri: "/auth/fetch",
        query: { sid: authCode },
        tag: "fetch-token"
      });
      return storageArea?.setItem(storageKey, JSON.stringify({ token })), state.set("setToken", { authState: { type: AuthStateType.LOGGED_IN, token, currentUser: null } }), cleanedUrl;
    } catch (error) {
      return state.set("exchangeSessionForTokenError", { authState: { type: AuthStateType.ERROR, error } }), cleanedUrl;
    }
  }
), logout = bindActionGlobally(authStore, async ({ state }) => {
  const { clientFactory, apiHost, providedToken, storageArea, storageKey } = state.get().options;
  if (providedToken) return;
  const { authState } = state.get();
  if (authState.type === AuthStateType.LOGGED_OUT && authState.isDestroyingSession) return;
  const token = authState.type === AuthStateType.LOGGED_IN && authState.token;
  try {
    token && (state.set("loggingOut", {
      authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !0 }
    }), await clientFactory({
      token,
      requestTagPrefix: REQUEST_TAG_PREFIX,
      apiVersion: DEFAULT_API_VERSION$1,
      ...apiHost && { apiHost },
      useProjectHostname: !1,
      useCdn: !1
    }).request({ uri: "/auth/logout", method: "POST" }));
  } finally {
    state.set("logoutSuccess", {
      authState: { type: AuthStateType.LOGGED_OUT, isDestroyingSession: !1 }
    }), storageArea?.removeItem(storageKey), storageArea?.removeItem(`${storageKey}_last_refresh`);
  }
}), destroyController$1 = ({ state }) => {
  const { controller } = state.get();
  controller && (controller.destroy(), state.set("destroyController", {
    controller: null,
    channels: /* @__PURE__ */ new Map()
  }));
}, getOrCreateChannel$1 = ({ state }, options) => {
  const controller = state.get().controller;
  if (!controller)
    throw new Error("Controller must be initialized before using or creating channels");
  const channels = state.get().channels, existing = channels.get(options.name);
  if (existing) {
    if (!isEqual(existing.options, options))
      throw new Error(`Channel "${options.name}" already exists with different options`);
    return state.set("incrementChannelRefCount", {
      channels: new Map(channels).set(options.name, {
        ...existing,
        refCount: existing.refCount + 1
      })
    }), existing.channel.start(), existing.channel;
  }
  const channel = controller.createChannel(options);
  return channel.start(), state.set("createChannel", {
    channels: new Map(channels).set(options.name, {
      channel,
      options,
      refCount: 1
    })
  }), channel;
}, getOrCreateController$1 = ({ state, instance }, targetOrigin) => {
  const { controller, controllerOrigin } = state.get();
  if (controller && controllerOrigin === targetOrigin)
    return controller;
  controller && destroyController$1({ state });
  const newController = createController({ targetOrigin });
  return state.set("initializeController", {
    controllerOrigin: targetOrigin,
    controller: newController
  }), newController;
}, releaseChannel$1 = ({ state }, name) => {
  const channels = state.get().channels, channelEntry = channels.get(name);
  if (channelEntry) {
    const newRefCount = channelEntry.refCount === 0 ? 0 : channelEntry.refCount - 1;
    newRefCount === 0 ? (channelEntry.channel.stop(), channels.delete(name), state.set("releaseChannel", { channels: new Map(channels) })) : state.set("releaseChannel", {
      channels: new Map(channels).set(name, {
        ...channelEntry,
        refCount: newRefCount
      })
    });
  }
}, comlinkControllerStore = {
  name: "connectionStore",
  getInitialState: () => ({
    controller: null,
    controllerOrigin: null,
    channels: /* @__PURE__ */ new Map()
  }),
  initialize({ instance }) {
    return () => {
      destroyController(instance);
    };
  }
}, destroyController = bindActionGlobally(
  comlinkControllerStore,
  destroyController$1
), getOrCreateChannel = bindActionGlobally(
  comlinkControllerStore,
  getOrCreateChannel$1
), getOrCreateController = bindActionGlobally(
  comlinkControllerStore,
  getOrCreateController$1
), releaseChannel = bindActionGlobally(comlinkControllerStore, releaseChannel$1), getOrCreateNode$1 = ({ state }, options) => {
  const nodes = state.get().nodes, existing = nodes.get(options.name);
  if (existing) {
    if (!isEqual(existing.options, options))
      throw new Error(`Node "${options.name}" already exists with different options`);
    return existing.node.start(), existing.node;
  }
  const node = createNode(options);
  node.start();
  const statusUnsub = node.onStatus((status) => {
    const currentNodes = state.get().nodes, currentEntry = currentNodes.get(options.name);
    if (!currentEntry) return;
    const updatedEntry = {
      ...currentEntry,
      status
    };
    state.set("updateNodeStatus", {
      nodes: new Map(currentNodes).set(options.name, updatedEntry)
    });
  }), entry = {
    node,
    options,
    status: "idle",
    statusUnsub
  };
  return nodes.set(options.name, entry), state.set("createNode", { nodes }), node;
}, releaseNode$1 = ({ state }, name) => {
  const nodes = state.get().nodes, existing = nodes.get(name);
  if (existing) {
    existing.statusUnsub && existing.statusUnsub(), existing.node.stop(), nodes.delete(name), state.set("removeNode", { nodes });
    return;
  }
}, comlinkNodeStore = {
  name: "nodeStore",
  getInitialState: () => ({
    nodes: /* @__PURE__ */ new Map(),
    subscriptions: /* @__PURE__ */ new Map()
  }),
  initialize({ state }) {
    return () => {
      state.get().nodes.forEach(({ node }) => {
        node.stop();
      });
    };
  }
}, releaseNode = bindActionGlobally(comlinkNodeStore, releaseNode$1), getOrCreateNode = bindActionGlobally(comlinkNodeStore, getOrCreateNode$1), NODE_RELEASE_TIME = 5e3, selectNode = (context, nodeInput) => context.state.nodes.get(nodeInput.name), getNodeState = bindActionGlobally(
  comlinkNodeStore,
  createStateSourceAction({
    selector: createSelector([selectNode], (nodeEntry) => nodeEntry?.status === "connected" ? {
      node: nodeEntry.node,
      status: nodeEntry.status
    } : void 0),
    onSubscribe: ({ state, instance }, nodeInput) => {
      const nodeName = nodeInput.name, subscriberId = Symbol("comlink-node-subscriber");
      getOrCreateNode(instance, nodeInput);
      let subs = state.get().subscriptions.get(nodeName);
      return subs || (subs = /* @__PURE__ */ new Set(), state.get().subscriptions.set(nodeName, subs)), subs.add(subscriberId), () => {
        setTimeout(() => {
          const activeSubs = state.get().subscriptions.get(nodeName);
          activeSubs && (activeSubs.delete(subscriberId), activeSubs.size === 0 && (state.get().subscriptions.delete(nodeName), releaseNode(instance, nodeName)));
        }, NODE_RELEASE_TIME);
      };
    }
  })
);
function createDocumentHandle(handle) {
  return handle;
}
function createDocumentTypeHandle(handle) {
  return handle;
}
function createProjectHandle(handle) {
  return handle;
}
function createDatasetHandle(handle) {
  return handle;
}
function configureLogging(config) {
  configureLogging$1(config);
  const configLevel = config.level || "warn", shouldLog = ["info", "debug", "trace"].includes(configLevel) || configLevel === "warn";
  shouldLog && config.handler?.info ? config.handler.info(`[${(/* @__PURE__ */ new Date()).toISOString()}] [INFO] [sdk] Logging configured`, {
    level: configLevel,
    namespaces: config.namespaces || [],
    internal: config.internal || !1,
    source: "programmatic"
  }) : shouldLog && console.info(`[${(/* @__PURE__ */ new Date()).toISOString()}] [INFO] [sdk] Logging configured`, {
    level: configLevel,
    namespaces: config.namespaces || [],
    internal: config.internal || !1,
    source: "programmatic"
  });
}
const API_VERSION$4 = "v2025-02-19", datasets = createFetcherStore({
  name: "Datasets",
  getKey: (instance, options) => {
    const projectId = options?.projectId ?? instance.config.projectId;
    if (!projectId)
      throw new Error("A projectId is required to use the project API.");
    return projectId;
  },
  fetcher: (instance) => (options) => getClientState(instance, {
    apiVersion: API_VERSION$4,
    // non-null assertion is fine because we check above
    projectId: options?.projectId ?? instance.config.projectId,
    useProjectHostname: !0
  }).observable.pipe(switchMap((client) => client.observable.datasets.list()))
}), getDatasetsState = datasets.getState, resolveDatasets = datasets.resolveState, isSanityMutatePatch = (value) => !(typeof value != "object" || !value || !("type" in value) || typeof value.type != "string" || value.type !== "patch" || !("id" in value) || typeof value.id != "string" || !("patches" in value) || !Array.isArray(value.patches));
function createDocument(doc, initialValue) {
  return {
    type: "document.create",
    ...doc,
    ...doc.documentId && {
      documentId: doc.liveEdit ? doc.documentId : getPublishedId(doc.documentId)
    },
    ...initialValue && { initialValue }
  };
}
function deleteDocument(doc) {
  return {
    type: "document.delete",
    ...doc,
    documentId: doc.liveEdit ? doc.documentId : getPublishedId(doc.documentId)
  };
}
function convertSanityMutatePatch(sanityPatchMutation) {
  return SanityEncoder.encode(sanityPatchMutation).map((i) => {
    const copy = { ...i.patch };
    return "id" in copy && delete copy.id, copy;
  });
}
function editDocument(doc, patches) {
  const documentId = doc.liveEdit ? doc.documentId : getPublishedId(doc.documentId);
  if (isSanityMutatePatch(patches)) {
    const converted = convertSanityMutatePatch(patches) ?? [];
    return {
      ...doc,
      type: "document.edit",
      documentId,
      patches: converted
    };
  }
  return {
    ...doc,
    type: "document.edit",
    documentId,
    ...patches && { patches: Array.isArray(patches) ? patches : [patches] }
  };
}
function publishDocument(doc) {
  return {
    type: "document.publish",
    ...doc,
    documentId: doc.liveEdit ? doc.documentId : getPublishedId(doc.documentId)
  };
}
function unpublishDocument(doc) {
  return {
    type: "document.unpublish",
    ...doc,
    documentId: doc.liveEdit ? doc.documentId : getPublishedId(doc.documentId)
  };
}
function discardDocument(doc) {
  return {
    type: "document.discard",
    ...doc,
    documentId: doc.liveEdit ? doc.documentId : getPublishedId(doc.documentId)
  };
}
const DOCUMENT_STATE_CLEAR_DELAY = 1e3, INITIAL_OUTGOING_THROTTLE_TIME = 1e3, API_VERSION$3 = "v2025-05-06";
function generateArrayKey(length = 12) {
  const numBytes = Math.ceil(length / 2), bytes = crypto.getRandomValues(new Uint8Array(numBytes));
  return Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("").slice(0, length);
}
function memoize(fn) {
  const cache = /* @__PURE__ */ new WeakMap();
  return ((input) => {
    if (!input || typeof input != "object") return fn(input);
    const cached = cache.get(input);
    if (cached) return cached;
    const result = fn(input);
    return cache.set(input, result), result;
  });
}
const ensureArrayKeysDeep = memoize((input) => {
  if (!input || typeof input != "object") return input;
  if (Array.isArray(input))
    return !input.length || typeof input[0] != "object" || input.every(isKeyedObject) ? input : input.map((item) => !item || typeof item != "object" ? item : isKeyedObject(item) ? ensureArrayKeysDeep(item) : { ...ensureArrayKeysDeep(item), _key: generateArrayKey() });
  const entries = Object.entries(input).map(([key, value]) => [key, ensureArrayKeysDeep(value)]);
  return entries.every(([key, value]) => input[key] === value) ? input : Object.fromEntries(entries);
});
function set(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(
    ([pathExpression, replacementValue]) => Array.from(jsonMatch(input, pathExpression)).map((matchEntry) => ({
      ...matchEntry,
      replacementValue
    }))
  ).reduce((acc, { path, replacementValue }) => setDeep(acc, path, replacementValue), input);
  return ensureArrayKeysDeep(result);
}
function setIfMissing(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(([pathExpression, replacementValue]) => Array.from(jsonMatch(input, pathExpression)).map((matchEntry) => ({
    ...matchEntry,
    replacementValue
  }))).filter((matchEntry) => matchEntry.value === null || matchEntry.value === void 0).reduce((acc, { path, replacementValue }) => setDeep(acc, path, replacementValue), input);
  return ensureArrayKeysDeep(result);
}
function unset(input, pathExpressions) {
  const result = pathExpressions.flatMap((pathExpression) => Array.from(jsonMatch(input, pathExpression))).reverse().reduce((acc, { path }) => unsetDeep(acc, path), input);
  return ensureArrayKeysDeep(result);
}
function insert(input, { items, ...insertPatch }) {
  let operation, pathExpression;
  if ("before" in insertPatch ? (operation = "before", pathExpression = insertPatch.before) : "after" in insertPatch ? (operation = "after", pathExpression = insertPatch.after) : "replace" in insertPatch && (operation = "replace", pathExpression = insertPatch.replace), !operation || typeof pathExpression != "string" || !pathExpression.length) return input;
  const arrayPath = slicePath(pathExpression, 0, -1), positionPath = slicePath(pathExpression, -1);
  let result = input;
  for (const { path, value } of jsonMatch(input, arrayPath)) {
    if (!Array.isArray(value)) continue;
    let arr = value;
    switch (operation) {
      case "replace": {
        const indexesToRemove = /* @__PURE__ */ new Set();
        let position = 1 / 0;
        for (const itemMatch of jsonMatch(arr, positionPath)) {
          if (itemMatch.path.length !== 1) continue;
          const [segment] = itemMatch.path;
          if (typeof segment == "string") continue;
          let index;
          typeof segment == "number" && (index = segment), typeof index == "number" && index < 0 && (index = arr.length + index), isKeySegment(segment) && (index = getIndexForKey(arr, segment._key)), typeof index == "number" && (index < 0 && (index = arr.length + index), indexesToRemove.add(index), index < position && (position = index));
        }
        if (position === 1 / 0) continue;
        arr = arr.map((item, index) => ({ item, index })).filter(({ index }) => !indexesToRemove.has(index)).map(({ item }) => item), arr = [...arr.slice(0, position), ...items, ...arr.slice(position, arr.length)];
        break;
      }
      case "before": {
        let position = 1 / 0;
        for (const itemMatch of jsonMatch(arr, positionPath)) {
          if (itemMatch.path.length !== 1) continue;
          const [segment] = itemMatch.path;
          if (typeof segment == "string") continue;
          let index;
          typeof segment == "number" && (index = segment), typeof index == "number" && index < 0 && (index = arr.length + index), isKeySegment(segment) && (index = getIndexForKey(arr, segment._key)), typeof index == "number" && (index < 0 && (index = arr.length - index), index < position && (position = index));
        }
        if (position === 1 / 0) continue;
        arr = [...arr.slice(0, position), ...items, ...arr.slice(position, arr.length)];
        break;
      }
      case "after": {
        let position = -1 / 0;
        for (const itemMatch of jsonMatch(arr, positionPath)) {
          if (itemMatch.path.length !== 1) continue;
          const [segment] = itemMatch.path;
          if (typeof segment == "string") continue;
          let index;
          typeof segment == "number" && (index = segment), typeof index == "number" && index < 0 && (index = arr.length + index), isKeySegment(segment) && (index = getIndexForKey(arr, segment._key)), typeof index == "number" && index > position && (position = index);
        }
        if (position === -1 / 0) continue;
        arr = [...arr.slice(0, position + 1), ...items, ...arr.slice(position + 1, arr.length)];
        break;
      }
      default:
        continue;
    }
    result = setDeep(result, path, arr);
  }
  return ensureArrayKeysDeep(result);
}
function inc(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(
    ([pathExpression, valueToAdd]) => Array.from(jsonMatch(input, pathExpression)).map((matchEntry) => ({
      ...matchEntry,
      valueToAdd
    }))
  ).filter(
    (matchEntry) => typeof matchEntry.value == "number"
  ).reduce((acc, { path, value, valueToAdd }) => setDeep(acc, path, value + valueToAdd), input);
  return ensureArrayKeysDeep(result);
}
function dec(input, pathExpressionValues) {
  const result = inc(
    input,
    Object.fromEntries(
      Object.entries(pathExpressionValues).filter(([, value]) => typeof value == "number").map(([key, value]) => [key, -value])
    )
  );
  return ensureArrayKeysDeep(result);
}
function diffMatchPatch(input, pathExpressionValues) {
  const result = Object.entries(pathExpressionValues).flatMap(
    ([pathExpression, dmp]) => Array.from(jsonMatch(input, pathExpression)).map((m) => ({ ...m, dmp }))
  ).filter((i) => i.value !== void 0).map(({ path, value, dmp }) => {
    if (typeof value != "string")
      throw new Error(
        `Can't diff-match-patch \`${JSON.stringify(value)}\` at path \`${stringifyPath(path)}\`, because it is not a string`
      );
    const [nextValue] = applyPatches(parsePatch(dmp), value);
    return { path, value: nextValue };
  }).reduce((acc, { path, value }) => setDeep(acc, path, value), input);
  return ensureArrayKeysDeep(result);
}
function ifRevisionID(input, revisionId) {
  const inputRev = typeof input == "object" && input && "_rev" in input && typeof input._rev == "string" ? input._rev : void 0;
  if (typeof inputRev != "string")
    throw new Error("Patch specified `ifRevisionID` but could not find document's revision ID.");
  if (revisionId !== inputRev)
    throw new Error(
      `Patch's \`ifRevisionID\` \`${revisionId}\` does not match document's revision ID \`${inputRev}\``
    );
  return input;
}
function setDeep(input, path, value) {
  const [currentSegment, ...restOfPath] = path;
  if (currentSegment === void 0) return value;
  if (typeof input != "object" || input === null) {
    if (typeof currentSegment == "string")
      return { [currentSegment]: setDeep(null, restOfPath, value) };
    let index;
    if (isKeySegment(currentSegment))
      index = 0;
    else if (typeof currentSegment == "number" && currentSegment >= 0)
      index = currentSegment;
    else
      return input;
    return [
      // fill until index
      ...Array.from({ length: index }).fill(null),
      // then set deep here
      setDeep(null, restOfPath, value)
    ];
  }
  if (Array.isArray(input)) {
    let index;
    return isKeySegment(currentSegment) ? index = getIndexForKey(input, currentSegment._key) ?? input.length : typeof currentSegment == "number" && (index = currentSegment < 0 ? input.length + currentSegment : currentSegment), index === void 0 ? input : index in input ? input.map(
      (nestedInput, i) => i === index ? setDeep(nestedInput, restOfPath, value) : nestedInput
    ) : [
      ...input,
      ...Array.from({ length: index - input.length }).fill(null),
      setDeep(null, restOfPath, value)
    ];
  }
  return typeof currentSegment == "object" ? input : currentSegment in input ? Object.fromEntries(
    Object.entries(input).map(
      ([key, nestedInput]) => key === currentSegment ? [key, setDeep(nestedInput, restOfPath, value)] : [key, nestedInput]
    )
  ) : { ...input, [currentSegment]: setDeep(null, restOfPath, value) };
}
function unsetDeep(input, path) {
  const [currentSegment, ...restOfPath] = path;
  if (currentSegment === void 0 || typeof input != "object" || input === null) return input;
  let _segment;
  if (isKeySegment(currentSegment) ? _segment = getIndexForKey(input, currentSegment._key) : (typeof currentSegment == "string" || typeof currentSegment == "number") && (_segment = currentSegment), _segment === void 0) return input;
  let segment = _segment;
  return typeof segment == "number" && Array.isArray(input) && (segment = segment < 0 ? input.length + segment : segment), segment in input ? restOfPath.length ? Array.isArray(input) ? input.map(
    (nestedInput, index) => index === segment ? unsetDeep(nestedInput, restOfPath) : nestedInput
  ) : Object.fromEntries(
    Object.entries(input).map(
      ([key, value]) => key === segment ? [key, unsetDeep(value, restOfPath)] : [key, value]
    )
  ) : Array.isArray(input) ? input.filter((_nestedInput, index) => index !== segment) : Object.fromEntries(Object.entries(input).filter(([key]) => key !== segment.toString())) : input;
}
const patchOperations = {
  ifRevisionID,
  set,
  setIfMissing,
  unset,
  inc,
  dec,
  insert,
  diffMatchPatch
};
function getId(id) {
  return !id || typeof id != "string" ? crypto.randomUUID() : id.endsWith(".") ? `${id}${crypto.randomUUID()}` : id;
}
function getDocumentIds(selection) {
  if ("id" in selection) {
    const ids = (Array.isArray(selection.id) ? selection.id : [selection.id]).filter((id) => typeof id == "string");
    return Array.from(new Set(ids));
  }
  if ("query" in selection)
    throw new Error("'query' in mutations is not supported.");
  return [];
}
function processMutations({
  documents,
  mutations,
  transactionId,
  timestamp
}) {
  if (!mutations.length) return documents;
  const dataset = { ...documents }, now = timestamp || (/* @__PURE__ */ new Date()).toISOString();
  for (const mutation of mutations) {
    if ("create" in mutation) {
      const id = getId(mutation.create._id);
      if (dataset[id])
        throw new Error(
          `Cannot create document with \`_id\` \`${id}\` because another document with the same ID already exists.`
        );
      const document2 = {
        // > `_createdAt` and `_updatedAt` may be submitted and will override
        // > the default which is of course the current time. This can be used
        // > to reconstruct a data-set with its timestamp structure intact.
        // >
        // > [- source](https://www.sanity.io/docs/http-mutations#c732f27330a4)
        _createdAt: now,
        _updatedAt: now,
        ...mutation.create,
        // prefer the user's `_createdAt` and `_updatedAt`
        _rev: transactionId,
        _id: id
      };
      dataset[id] = document2;
      continue;
    }
    if ("createOrReplace" in mutation) {
      const id = getId(mutation.createOrReplace._id), prev = dataset[id], document2 = {
        ...mutation.createOrReplace,
        // otherwise, if the mutation provided, a `_createdAt` time, use it,
        // otherwise default to now
        _createdAt: (
          // if there was an existing document, use the previous `_createdAt`
          // since we're replacing the current document
          prev?._createdAt || // if there was no previous document, then we're creating this
          // document for the first time so we should use the `_createdAt` from
          // the mutation if the user included it
          typeof mutation.createOrReplace._createdAt == "string" && mutation.createOrReplace._createdAt || // otherwise, default to now
          now
        ),
        _updatedAt: (
          // if there was an existing document, then set the `_updatedAt` to now
          // since we're replacing the current document
          prev ? now : (
            // otherwise, we're creating this document for the first time,
            // in that case, use the user's `_updatedAt` if included in the
            // mutation
            typeof mutation.createOrReplace._updatedAt == "string" && mutation.createOrReplace._updatedAt || // otherwise default to now
            now
          )
        ),
        _rev: transactionId,
        _id: id
      };
      dataset[id] = document2;
      continue;
    }
    if ("createIfNotExists" in mutation) {
      const id = getId(mutation.createIfNotExists._id);
      if (dataset[id]) continue;
      const document2 = {
        // same logic as `create`:
        // prefer the user's `_createdAt` and `_updatedAt`
        _createdAt: now,
        _updatedAt: now,
        ...mutation.createIfNotExists,
        _rev: transactionId,
        _id: id
      };
      dataset[id] = document2;
      continue;
    }
    if ("delete" in mutation) {
      for (const id of getDocumentIds(mutation.delete))
        dataset[id] = null;
      continue;
    }
    if ("patch" in mutation) {
      const { patch } = mutation, patched = getDocumentIds(patch).map((id) => {
        if (!dataset[id])
          throw new Error(`Cannot patch document with ID \`${id}\` because it was not found.`);
        return Object.entries(patchOperations).reduce((acc, [type, operation]) => patch[type] ? operation(
          acc,
          // @ts-expect-error TS doesn't handle this union very well
          patch[type]
        ) : acc, dataset[id]);
      });
      for (const result of patched)
        dataset[result._id] = {
          ...result,
          _rev: transactionId,
          _updatedAt: now
        };
      continue;
    }
  }
  return dataset;
}
const DEFAULT_MAX_BUFFER_SIZE = 20, DEFAULT_DEADLINE_MS = 3e4;
class OutOfSyncError extends Error {
  /**
   * Attach state to the error for debugging/reporting
   */
  state;
  constructor(message, state) {
    super(message), this.name = "OutOfSyncError", this.state = state;
  }
}
class DeadlineExceededError extends OutOfSyncError {
  constructor(message, state) {
    super(message, state), this.name = "DeadlineExceededError";
  }
}
class MaxBufferExceededError extends OutOfSyncError {
  constructor(message, state) {
    super(message, state), this.name = "MaxBufferExceededError";
  }
}
function sortListenerEvents(options) {
  const { resolveChainDeadline = DEFAULT_DEADLINE_MS, maxBufferSize = DEFAULT_MAX_BUFFER_SIZE } = {};
  return (input$) => input$.pipe(
    // Maintain state: current base revision, a buffer of pending mutation events,
    // and a list of events to emit.
    scan(
      (state, event) => {
        if (event.type === "sync")
          return {
            base: { revision: event.document?._rev },
            buffer: [],
            emitEvents: [event]
          };
        if (event.type === "mutation") {
          if (!state.base)
            throw new Error(
              "Invalid state. Cannot process mutation event without a base sync event"
            );
          const buffer = state.buffer.concat(event), emitEvents = [];
          let baseRevision = state.base.revision, progress = !0;
          for (; progress; ) {
            progress = !1;
            const idx = buffer.findIndex((e) => e.previousRev === baseRevision);
            if (idx !== -1) {
              const [next] = buffer.splice(idx, 1);
              emitEvents.push(next), baseRevision = next.transition === "disappear" ? void 0 : next.resultRev, progress = !0;
            }
          }
          if (buffer.length >= maxBufferSize)
            throw new MaxBufferExceededError(
              `Too many unchainable mutation events (${buffer.length}) waiting to resolve.`,
              { base: { revision: baseRevision }, buffer, emitEvents }
            );
          return {
            base: { revision: baseRevision },
            buffer,
            emitEvents
          };
        }
        return { ...state, emitEvents: [event] };
      },
      {
        base: void 0,
        buffer: [],
        emitEvents: []
      }
    ),
    switchMap((state) => state.buffer.length > 0 ? concat(
      of(state),
      timer(resolveChainDeadline).pipe(
        mergeMap(
          () => throwError(
            () => new DeadlineExceededError(
              `Did not resolve chain within a deadline of ${resolveChainDeadline}ms`,
              state
            )
          )
        )
      )
    ) : of(state)),
    // Emit all events that are ready to be applied.
    mergeMap((state) => of(...state.emitEvents))
  );
}
const listen = ({ state }, documentId) => {
  const { sharedListener, fetchDocument } = state.get();
  return sharedListener.events.pipe(
    concatMap((e) => e.type === "welcome" ? fetchDocument(documentId).pipe(
      map((document2) => ({ type: "sync", document: document2 }))
    ) : e.type === "mutation" && e.documentId === documentId ? of(e) : EMPTY),
    sortListenerEvents(),
    withLatestFrom(
      state.observable.pipe(
        map((s) => s.documentStates[documentId]),
        filter(Boolean),
        distinctUntilChanged()
      )
    ),
    map(([next, documentState]) => {
      if (next.type === "sync")
        return {
          type: "sync",
          documentId,
          document: next.document,
          revision: next.document?._rev,
          timestamp: next.document?._updatedAt ?? (/* @__PURE__ */ new Date()).toISOString()
        };
      const [document2] = Object.values(
        processMutations({
          documents: { [documentId]: documentState.remote },
          mutations: next.mutations,
          transactionId: next.transactionId,
          timestamp: next.timestamp
        })
      ), { previousRev, transactionId, timestamp } = next;
      return {
        type: "mutation",
        documentId,
        document: document2 ?? null,
        revision: transactionId,
        timestamp,
        ...previousRev && { previousRev }
      };
    })
  );
};
class MultiKeyWeakMap {
  // The root of our nested WeakMap structure.
  #rootMap = /* @__PURE__ */ new WeakMap();
  // Instead of random IDs, we use a counter for deterministic IDs.
  static #globalIdCounter = 0;
  // Each instance keeps a cache mapping a key to its assigned ID.
  #idCache = /* @__PURE__ */ new WeakMap();
  /**
   * Assigns a numeric ID to the key.
   */
  #assignId(key) {
    const cachedId = this.#idCache.get(key);
    if (cachedId !== void 0) return cachedId;
    const id = MultiKeyWeakMap.#globalIdCounter;
    return this.#idCache.set(key, id), MultiKeyWeakMap.#globalIdCounter++, id;
  }
  /**
   * Remove duplicate keys and arrange them in a consistent order
   * by sorting according to their assigned IDs.
   */
  #arrangeKeys(keys) {
    const keyed = Array.from(new Set(keys)).map((key) => [this.#assignId(key), key]);
    return keyed.sort((a, b) => a[0] - b[0]), keyed.map(([, key]) => key);
  }
  /**
   * Recursively search the nested WeakMap structure for the value.
   */
  #getDeep(keys, map2) {
    if (keys.length === 0) return;
    const [firstKey, ...restKeys] = keys, node = map2.get(firstKey);
    if (node)
      return restKeys.length === 0 ? node.value : this.#getDeep(restKeys, node.next);
  }
  /**
   * Recursively create nodes along the key chain until the final key
   * is reached, then assign the value.
   */
  #setDeep(keys, map2, value) {
    if (keys.length === 0) return;
    const [firstKey, ...restKeys] = keys;
    let node = map2.get(firstKey);
    node || (node = {
      value: void 0,
      next: /* @__PURE__ */ new WeakMap()
    }, map2.set(firstKey, node)), restKeys.length === 0 ? node.value = value : this.#setDeep(restKeys, node.next, value);
  }
  /**
   * Retrieves the value associated with the array of keys.
   * The keys are de-duplicated and sorted so that the order does not matter.
   */
  get(keys) {
    const arrangedKeys = this.#arrangeKeys(keys);
    return this.#getDeep(arrangedKeys, this.#rootMap);
  }
  /**
   * Associates the value with the given array of keys.
   */
  set(keys, value) {
    const arrangedKeys = this.#arrangeKeys(keys);
    this.#setDeep(arrangedKeys, this.#rootMap, value);
  }
}
function createGrantsLookup(datasetAcl) {
  const filtersByGrant = {
    create: /* @__PURE__ */ new Set(),
    history: /* @__PURE__ */ new Set(),
    read: /* @__PURE__ */ new Set(),
    update: /* @__PURE__ */ new Set()
  };
  for (const entry of datasetAcl)
    for (const grant of entry.permissions) {
      const set2 = filtersByGrant[grant];
      set2.add(entry.filter), filtersByGrant[grant] = set2;
    }
  return Object.fromEntries(
    Object.entries(filtersByGrant).map(([grant, filters]) => {
      const combinedFilter = Array.from(filters).map((i) => `(${i})`).join("||");
      return combinedFilter ? [grant, parse(`$document {"_": ${combinedFilter}}._`)] : [grant, parse("false")];
    })
  );
}
const documentsCache = new MultiKeyWeakMap(), actionsCache = /* @__PURE__ */ new WeakMap(), nullReplacer = {}, documentsSelector = createSelector(
  [
    ({ state: { documentStates } }) => documentStates,
    (_context, { actions }) => actions
  ],
  (documentStates, actions) => {
    const documentIds = new Set(
      actions.map((action) => typeof action.documentId != "string" ? [] : action.liveEdit ? [action.documentId] : [getPublishedId(action.documentId), getDraftId(action.documentId)]).flat()
    ), documents = {};
    for (const documentId of documentIds) {
      const local = documentStates[documentId]?.local;
      if (local === void 0) return;
      documents[documentId] = local;
    }
    const keys = Object.values(
      // value in this record will be `undefined` because
      // of the early return if undefined is found above
      documents
    ).map((doc) => doc === null ? nullReplacer : doc);
    return documentsCache.get(keys) || (documentsCache.set(keys, documents), documents);
  }
), memoizedActionsSelector = createSelector(
  [
    documentsSelector,
    (_state, { actions }) => actions
  ],
  (documents, actions) => {
    if (!documents) return;
    let nestedCache = actionsCache.get(documents);
    nestedCache || (nestedCache = /* @__PURE__ */ new Map(), actionsCache.set(documents, nestedCache));
    const normalizedActions = Array.isArray(actions) ? actions : [actions], actionsKey = JSON.stringify(normalizedActions);
    return nestedCache.get(actionsKey) || (nestedCache.set(actionsKey, normalizedActions), normalizedActions);
  }
);
function checkGrant$1(grantExpr, document2) {
  const value = evaluateSync(grantExpr, { params: { document: document2 } });
  return value.type === "boolean" && value.data;
}
const enNarrowConjunction = new Intl.ListFormat("en", { style: "narrow", type: "conjunction" });
function calculatePermissions(...args) {
  return _calculatePermissions(...args);
}
const _calculatePermissions = createSelector(
  [
    ({ state: { grants } }) => grants,
    documentsSelector,
    memoizedActionsSelector
  ],
  (grants, documents, actions) => {
    if (!documents || !grants || !actions) return;
    const timestamp = (/* @__PURE__ */ new Date()).toISOString(), reasons = [];
    try {
      processActions({
        actions,
        transactionId: crypto.randomUUID(),
        working: documents,
        base: documents,
        timestamp,
        grants
      });
    } catch (error) {
      if (error instanceof PermissionActionError)
        reasons.push({
          message: error.message,
          documentId: error.documentId,
          type: "access"
        });
      else if (error instanceof ActionError)
        reasons.push({
          message: error.message,
          documentId: error.documentId,
          type: "precondition"
        });
      else
        throw error;
    }
    for (const action of actions)
      if (action.type === "document.edit" && !action.patches?.length) {
        const docId = action.documentId, doc = action.liveEdit ? documents[docId] : documents[getDraftId(docId)] ?? documents[getPublishedId(docId)];
        doc ? checkGrant$1(grants.update, doc) || reasons.push({
          type: "access",
          message: `You are not allowed to edit the document with ID "${docId}".`,
          documentId: docId
        }) : reasons.push({
          type: "precondition",
          message: `The document with ID "${docId}" could not be found. Please check that it exists before editing.`,
          documentId: docId
        });
      }
    const allowed = reasons.length === 0;
    if (allowed) return { allowed };
    const sortedReasons = reasons.map((reason, index) => ({ ...reason, index })).sort((a, b) => a.type !== b.type ? a.type === "access" ? -1 : 1 : a.message.localeCompare(b.message, "en-US")).map(({ index: _index, ...reason }) => reason);
    return {
      allowed,
      reasons: sortedReasons,
      message: enNarrowConjunction.format(sortedReasons.map((i) => i.message))
    };
  }
);
function checkGrant(grantExpr, document2) {
  const value = evaluateSync(grantExpr, { params: { document: document2 } });
  return value.type === "boolean" && value.data;
}
class ActionError extends Error {
  documentId;
  transactionId;
  constructor(options) {
    super(options.message), Object.assign(this, options);
  }
}
class PermissionActionError extends ActionError {
}
function processActions({
  actions,
  transactionId,
  working: initialWorking,
  base: initialBase,
  timestamp,
  grants
}) {
  let working = { ...initialWorking }, base = { ...initialBase };
  const outgoingActions = [], outgoingMutations = [];
  for (const action of actions)
    switch (action.type) {
      case "document.create": {
        const documentId = getId(action.documentId);
        if (action.liveEdit) {
          if (working[documentId])
            throw new ActionError({
              documentId,
              transactionId,
              message: "This document already exists."
            });
          const newDocBase2 = { _type: action.documentType, _id: documentId }, newDocWorking2 = { _type: action.documentType, _id: documentId }, mutations2 = [{ create: newDocWorking2 }];
          if (base = processMutations({
            documents: base,
            transactionId,
            mutations: [{ create: newDocBase2 }],
            timestamp
          }), working = processMutations({
            documents: working,
            transactionId,
            mutations: mutations2,
            timestamp
          }), !checkGrant(grants.create, working[documentId]))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: `You do not have permission to create document "${documentId}".`
            });
          outgoingMutations.push(...mutations2), outgoingActions.push({
            actionType: "sanity.action.document.create",
            publishedId: documentId,
            attributes: newDocWorking2
          });
          continue;
        }
        const draftId = getDraftId(documentId), publishedId = getPublishedId(documentId);
        if (working[draftId])
          throw new ActionError({
            documentId,
            transactionId,
            message: "A draft version of this document already exists. Please use or discard the existing draft before creating a new one."
          });
        const newDocBase = {
          ...base[publishedId],
          _type: action.documentType,
          _id: draftId,
          ...action.initialValue
        }, newDocWorking = {
          ...working[publishedId],
          _type: action.documentType,
          _id: draftId,
          ...action.initialValue
        }, mutations = [{ create: newDocWorking }];
        if (base = processMutations({
          documents: base,
          transactionId,
          mutations: [{ create: newDocBase }],
          timestamp
        }), working = processMutations({
          documents: working,
          transactionId,
          mutations,
          timestamp
        }), !checkGrant(grants.create, working[draftId]))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to create a draft for document "${documentId}".`
          });
        outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.version.create",
          publishedId,
          attributes: newDocWorking
        });
        continue;
      }
      case "document.delete": {
        const documentId = action.documentId;
        if (action.liveEdit) {
          if (!working[documentId])
            throw new ActionError({
              documentId,
              transactionId,
              message: "The document you are trying to delete does not exist."
            });
          if (!checkGrant(grants.update, working[documentId]))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: "You do not have permission to delete this document."
            });
          const mutations2 = [{ delete: { id: documentId } }];
          base = processMutations({ documents: base, transactionId, mutations: mutations2, timestamp }), working = processMutations({ documents: working, transactionId, mutations: mutations2, timestamp }), outgoingMutations.push(...mutations2), outgoingActions.push({
            actionType: "sanity.action.document.delete",
            publishedId: documentId
          });
          continue;
        }
        const draftId = getDraftId(documentId), publishedId = getPublishedId(documentId);
        if (!working[publishedId])
          throw new ActionError({
            documentId,
            transactionId,
            message: working[draftId] ? "Cannot delete a document without a published version." : "The document you are trying to delete does not exist."
          });
        const cantDeleteDraft = working[draftId] && !checkGrant(grants.update, working[draftId]), cantDeletePublished = working[publishedId] && !checkGrant(grants.update, working[publishedId]);
        if (cantDeleteDraft || cantDeletePublished)
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: "You do not have permission to delete this document."
          });
        const mutations = [{ delete: { id: publishedId } }, { delete: { id: draftId } }], includeDrafts = working[draftId] ? [draftId] : void 0;
        base = processMutations({ documents: base, transactionId, mutations, timestamp }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.delete",
          publishedId,
          ...includeDrafts ? { includeDrafts } : {}
        });
        continue;
      }
      case "document.discard": {
        const documentId = getId(action.documentId);
        if (action.liveEdit)
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot discard changes for liveEdit document "${documentId}". LiveEdit documents do not support drafts.`
          });
        const draftId = getDraftId(documentId), mutations = [{ delete: { id: draftId } }];
        if (!working[draftId])
          throw new ActionError({
            documentId,
            transactionId,
            message: `There is no draft available to discard for document "${documentId}".`
          });
        if (!checkGrant(grants.update, working[draftId]))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to discard changes for document "${documentId}".`
          });
        base = processMutations({ documents: base, transactionId, mutations, timestamp }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.version.discard",
          versionId: draftId
        });
        continue;
      }
      case "document.edit": {
        const documentId = getId(action.documentId);
        if (action.liveEdit) {
          const userPatches2 = action.patches?.map((patch) => ({ patch: { id: documentId, ...patch } }));
          if (!userPatches2?.length) continue;
          if (!working[documentId] || !base[documentId])
            throw new ActionError({
              documentId,
              transactionId,
              message: "Cannot edit document because it does not exist."
            });
          const baseBefore2 = base[documentId];
          userPatches2 && (base = processMutations({
            documents: base,
            transactionId,
            mutations: userPatches2,
            timestamp
          }));
          const baseAfter2 = base[documentId], patches2 = diffValue(baseBefore2, baseAfter2), workingBefore2 = working[documentId];
          if (!checkGrant(grants.update, workingBefore2))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: `You do not have permission to edit document "${documentId}".`
            });
          const workingMutations2 = patches2.map((patch) => ({ patch: { id: documentId, ...patch } }));
          working = processMutations({
            documents: working,
            transactionId,
            mutations: workingMutations2,
            timestamp
          }), outgoingMutations.push(...workingMutations2), outgoingActions.push(
            ...patches2.map(
              (patch) => ({
                actionType: "sanity.action.document.edit",
                // Server requires draftId to have drafts. prefix for validation, even for liveEdit
                draftId: getDraftId(documentId),
                publishedId: documentId,
                patch
              })
            )
          );
          continue;
        }
        const draftId = getDraftId(documentId), publishedId = getPublishedId(documentId), userPatches = action.patches?.map((patch) => ({ patch: { id: draftId, ...patch } }));
        if (!userPatches?.length) continue;
        if (!working[draftId] && !working[publishedId] || !base[draftId] && !base[publishedId])
          throw new ActionError({
            documentId,
            transactionId,
            message: "Cannot edit document because it does not exist in draft or published form."
          });
        const baseMutations = [];
        !base[draftId] && base[publishedId] && baseMutations.push({ create: { ...base[publishedId], _id: draftId } });
        const baseBefore = base[draftId] ?? base[publishedId];
        userPatches && baseMutations.push(...userPatches), base = processMutations({
          documents: base,
          transactionId,
          mutations: baseMutations,
          timestamp
        });
        const baseAfter = base[draftId], patches = diffValue(baseBefore, baseAfter), workingMutations = [];
        if (!working[draftId] && working[publishedId]) {
          const newDraftFromPublished = { ...working[publishedId], _id: draftId };
          if (!checkGrant(grants.create, newDraftFromPublished))
            throw new PermissionActionError({
              documentId,
              transactionId,
              message: "You do not have permission to create a draft for editing this document."
            });
          workingMutations.push({ create: newDraftFromPublished });
        }
        const workingBefore = working[draftId] ?? working[publishedId];
        if (!checkGrant(grants.update, workingBefore))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to edit document "${documentId}".`
          });
        workingMutations.push(...patches.map((patch) => ({ patch: { id: draftId, ...patch } }))), working = processMutations({
          documents: working,
          transactionId,
          mutations: workingMutations,
          timestamp
        }), outgoingMutations.push(...workingMutations), outgoingActions.push(
          ...patches.map(
            (patch) => ({
              actionType: "sanity.action.document.edit",
              draftId,
              publishedId,
              patch
            })
          )
        );
        continue;
      }
      case "document.publish": {
        const documentId = getId(action.documentId);
        if (action.liveEdit)
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot publish liveEdit document "${documentId}". LiveEdit documents do not support drafts or publishing.`
          });
        const draftId = getDraftId(documentId), publishedId = getPublishedId(documentId), workingDraft = working[draftId], baseDraft = base[draftId];
        if (!workingDraft || !baseDraft)
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot publish because no draft version was found for document "${documentId}".`
          });
        if (!isEqual(workingDraft, baseDraft))
          throw new ActionError({
            documentId,
            transactionId,
            message: "Publish aborted: The document has changed elsewhere. Please try again."
          });
        const newPublishedFromDraft = { ...strengthenOnPublish(workingDraft), _id: publishedId }, mutations = [
          { delete: { id: draftId } },
          { createOrReplace: newPublishedFromDraft }
        ];
        if (working[draftId] && !checkGrant(grants.update, working[draftId]))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `Publish failed: You do not have permission to update the draft for "${documentId}".`
          });
        if (working[publishedId] && !checkGrant(grants.update, newPublishedFromDraft))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `Publish failed: You do not have permission to update the published version of "${documentId}".`
          });
        if (!working[publishedId] && !checkGrant(grants.create, newPublishedFromDraft))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `Publish failed: You do not have permission to publish a new version of "${documentId}".`
          });
        base = processMutations({ documents: base, transactionId, mutations, timestamp }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.publish",
          draftId,
          publishedId
        });
        continue;
      }
      case "document.unpublish": {
        const documentId = getId(action.documentId);
        if (action.liveEdit)
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot unpublish liveEdit document "${documentId}". LiveEdit documents do not support drafts or publishing.`
          });
        const draftId = getDraftId(documentId), publishedId = getPublishedId(documentId);
        if (!working[publishedId] && !base[publishedId])
          throw new ActionError({
            documentId,
            transactionId,
            message: `Cannot unpublish because the document "${documentId}" is not currently published.`
          });
        const sourceDoc = working[publishedId] ?? base[publishedId], newDraftFromPublished = { ...sourceDoc, _id: draftId }, mutations = [
          { delete: { id: publishedId } },
          { createIfNotExists: newDraftFromPublished }
        ];
        if (!checkGrant(grants.update, sourceDoc))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to unpublish the document "${documentId}".`
          });
        if (!working[draftId] && !checkGrant(grants.create, newDraftFromPublished))
          throw new PermissionActionError({
            documentId,
            transactionId,
            message: `You do not have permission to create a draft from the published version of "${documentId}".`
          });
        base = processMutations({
          documents: base,
          transactionId,
          mutations: [
            { delete: { id: publishedId } },
            { createIfNotExists: { ...base[publishedId] ?? sourceDoc, _id: draftId } }
          ],
          timestamp
        }), working = processMutations({ documents: working, transactionId, mutations, timestamp }), outgoingMutations.push(...mutations), outgoingActions.push({
          actionType: "sanity.action.document.unpublish",
          draftId,
          publishedId
        });
        continue;
      }
      default:
        throw new Error(
          `Unknown action type: "${// @ts-expect-error invalid input
          action.type}". Please contact support if this issue persists.`
        );
    }
  const previousRevs = Object.fromEntries(
    Object.entries(initialWorking).map(([id, doc]) => [id, doc?._rev])
  );
  return {
    working,
    outgoingActions,
    outgoingMutations,
    previous: initialWorking,
    previousRevs
  };
}
function strengthenOnPublish(draft) {
  const isStrengthenReference = (value) => "_strengthenOnPublish" in value;
  function strengthen(value) {
    if (typeof value != "object" || !value) return value;
    if (isStrengthenReference(value)) {
      const { _strengthenOnPublish, _weak, ...rest } = value;
      return {
        ...rest,
        ..._strengthenOnPublish.weak && { _weak: !0 }
      };
    }
    return Array.isArray(value) ? value.map(strengthen) : Object.fromEntries(Object.entries(value).map(([k, v]) => [k, strengthen(v)]));
  }
  return strengthen(draft);
}
const EMPTY_REVISIONS = {};
function queueTransaction(prev, transaction) {
  const { transactionId, actions } = transaction;
  return {
    ...getDocumentIdsFromActions(actions).reduce(
      (acc, id) => addSubscriptionIdToDocument(acc, id, transactionId),
      prev
    ),
    queued: [...prev.queued, transaction]
  };
}
function removeQueuedTransaction(prev, transactionId) {
  const transaction = prev.queued.find((t) => t.transactionId === transactionId);
  return transaction ? {
    ...getDocumentIdsFromActions(transaction.actions).reduce(
      (acc, id) => removeSubscriptionIdFromDocument(acc, id, transactionId),
      prev
    ),
    queued: prev.queued.filter((t) => transactionId !== t.transactionId)
  } : prev;
}
function applyFirstQueuedTransaction(prev) {
  const queued = prev.queued.at(0);
  if (!queued || !prev.grants) return prev;
  const ids = getDocumentIdsFromActions(queued.actions);
  if (ids.some((id) => prev.documentStates[id]?.local === void 0)) return prev;
  const working = ids.reduce((acc, id) => (acc[id] = prev.documentStates[id]?.local, acc), {}), timestamp = (/* @__PURE__ */ new Date()).toISOString(), result = processActions({
    ...queued,
    working,
    base: working,
    timestamp,
    grants: prev.grants
  }), applied = {
    ...queued,
    ...result,
    base: result.previous,
    timestamp
  };
  return {
    ...prev,
    applied: [...prev.applied, applied],
    queued: prev.queued.filter((t) => t.transactionId !== queued.transactionId),
    documentStates: Object.entries(result.working).reduce(
      (acc, [id, next]) => {
        const prevDoc = acc[id];
        return prevDoc && (acc[id] = { ...prevDoc, local: next }), acc;
      },
      { ...prev.documentStates }
    )
  };
}
function batchAppliedTransactions([curr, ...rest]) {
  if (!curr) return;
  if (!curr.actions.length) return batchAppliedTransactions(rest);
  if (curr.actions.length > 1)
    return {
      ...curr,
      disableBatching: !0,
      batchedTransactionIds: [curr.transactionId]
    };
  const [action] = curr.actions;
  if (action.type !== "document.edit" || curr.disableBatching)
    return {
      ...curr,
      disableBatching: !0,
      batchedTransactionIds: [curr.transactionId]
    };
  const editAction = {
    ...curr,
    actions: [action],
    disableBatching: !1,
    batchedTransactionIds: [curr.transactionId]
  };
  if (!rest.length) return editAction;
  const next = batchAppliedTransactions(rest);
  if (next)
    return next.disableBatching ? editAction : {
      disableBatching: !1,
      // Use the transactionId from the later (next) transaction.
      transactionId: next.transactionId,
      // Accumulate actions: current action first, then later ones.
      actions: [action, ...next.actions],
      // Merge outgoingActions in order.
      outgoingActions: [...curr.outgoingActions, ...next.outgoingActions],
      // Batched transaction IDs: preserve order by placing curr first.
      batchedTransactionIds: [curr.transactionId, ...next.batchedTransactionIds],
      // Merge outgoingMutations in order.
      outgoingMutations: [...curr.outgoingMutations, ...next.outgoingMutations],
      // Working state reflects the latest optimistic changes: later transactions override earlier.
      working: { ...curr.working, ...next.working },
      // Base state (base, previous, previousRevs) must reflect the original state.
      // Use curr values (the earliest transaction) to override later ones.
      previousRevs: { ...next.previousRevs, ...curr.previousRevs },
      previous: { ...next.previous, ...curr.previous },
      base: { ...next.base, ...curr.base },
      // Use the earliest timestamp from curr.
      timestamp: curr.timestamp ?? next.timestamp
    };
}
function transitionAppliedTransactionsToOutgoing(prev) {
  if (prev.outgoing) return prev;
  const transaction = batchAppliedTransactions(prev.applied);
  if (!transaction) return prev;
  const {
    transactionId,
    previousRevs,
    working,
    batchedTransactionIds: consumedTransactions
  } = transaction, timestamp = (/* @__PURE__ */ new Date()).toISOString();
  return {
    ...prev,
    outgoing: transaction,
    applied: prev.applied.filter((i) => !consumedTransactions.includes(i.transactionId)),
    documentStates: Object.entries(previousRevs).reduce(
      (acc, [documentId, previousRev]) => {
        if (working[documentId]?._rev === previousRev) return acc;
        const documentState = prev.documentStates[documentId];
        return documentState && (acc[documentId] = {
          ...documentState,
          unverifiedRevisions: {
            ...documentState.unverifiedRevisions,
            // add unverified revision
            [transactionId]: { documentId, previousRev, transactionId, timestamp }
          }
        }), acc;
      },
      { ...prev.documentStates }
    )
  };
}
function cleanupOutgoingTransaction(prev) {
  const { outgoing } = prev;
  if (!outgoing) return prev;
  let next = prev;
  const ids = getDocumentIdsFromActions(outgoing.actions);
  for (const transactionId of outgoing.batchedTransactionIds)
    for (const documentId of ids)
      next = removeSubscriptionIdFromDocument(next, documentId, transactionId);
  return { ...next, outgoing: void 0 };
}
function revertOutgoingTransaction(prev) {
  if (!prev.grants) return prev;
  let working = Object.fromEntries(
    Object.entries(prev.documentStates).map(([documentId, documentState]) => [
      documentId,
      documentState?.remote
    ])
  );
  const nextApplied = [];
  for (const t of prev.applied)
    try {
      const next = processActions({ ...t, working, grants: prev.grants });
      working = next.working, nextApplied.push({ ...t, ...next });
    } catch (error) {
      if (error instanceof ActionError) continue;
      throw error;
    }
  return {
    ...prev,
    applied: nextApplied,
    outgoing: void 0,
    documentStates: Object.fromEntries(
      Object.entries(prev.documentStates).filter((e) => !!e[1]).map(([documentId, { unverifiedRevisions = {}, local, ...documentState }]) => {
        const next = {
          ...documentState,
          local: documentId in working ? working[documentId] : local,
          unverifiedRevisions: prev.outgoing && prev.outgoing.transactionId in unverifiedRevisions ? omit(unverifiedRevisions, prev.outgoing.transactionId) : unverifiedRevisions
        };
        return [documentId, next];
      })
    )
  };
}
function applyRemoteDocument(prev, { document: document2, documentId, previousRev, revision, timestamp, type }, events) {
  if (!prev.grants) return prev;
  const prevDocState = prev.documentStates[documentId];
  if (!prevDocState) return prev;
  const prevUnverifiedRevisions = prevDocState.unverifiedRevisions, revisionToVerify = revision ? prevUnverifiedRevisions?.[revision] : void 0;
  let unverifiedRevisions = prevUnverifiedRevisions ?? EMPTY_REVISIONS;
  if (revision && revisionToVerify && (unverifiedRevisions = omit(prevUnverifiedRevisions, revision)), type === "sync" && (unverifiedRevisions = Object.fromEntries(
    Object.entries(unverifiedRevisions).filter(([, unverifiedRevision]) => unverifiedRevision ? new Date(timestamp).getTime() <= new Date(unverifiedRevision.timestamp).getTime() : !1)
  )), revisionToVerify && revisionToVerify.previousRev === previousRev)
    return {
      ...prev,
      documentStates: {
        ...prev.documentStates,
        [documentId]: {
          ...prevDocState,
          remote: document2,
          remoteRev: revision,
          unverifiedRevisions
        }
      }
    };
  let working = { ...prev.applied.at(0)?.previous, [documentId]: document2 };
  const nextApplied = [];
  for (const curr of prev.applied)
    try {
      const next = processActions({ ...curr, working, grants: prev.grants });
      working = next.working, nextApplied.push({ ...curr, ...next });
    } catch (error) {
      if (error instanceof ActionError) {
        events.next({
          type: "rebase-error",
          transactionId: error.transactionId,
          documentId: error.documentId,
          message: error.message,
          error
        });
        continue;
      }
      throw error;
    }
  return {
    ...prev,
    applied: nextApplied,
    documentStates: {
      ...prev.documentStates,
      [documentId]: {
        ...prevDocState,
        remote: document2,
        remoteRev: revision,
        local: working[documentId],
        unverifiedRevisions
      }
    }
  };
}
function addSubscriptionIdToDocument(prev, documentId, subscriptionId) {
  const prevDocState = prev.documentStates?.[documentId], prevSubscriptions = prevDocState?.subscriptions ?? [];
  return {
    ...prev,
    documentStates: {
      ...prev.documentStates,
      [documentId]: {
        ...prevDocState,
        id: documentId,
        subscriptions: [...prevSubscriptions, subscriptionId]
      }
    }
  };
}
function removeSubscriptionIdFromDocument(prev, documentId, subscriptionId) {
  const prevDocState = prev.documentStates?.[documentId], subscriptions = (prevDocState?.subscriptions ?? []).filter((id) => id !== subscriptionId);
  return prevDocState ? subscriptions.length ? {
    ...prev,
    documentStates: {
      ...prev.documentStates,
      [documentId]: { ...prevDocState, subscriptions }
    }
  } : { ...prev, documentStates: omit(prev.documentStates, documentId) } : prev;
}
function manageSubscriberIds({ state }, documentId, options) {
  const expandDraftPublished = options?.expandDraftPublished ?? !0, documentIds = Array.from(
    new Set(
      expandDraftPublished ? (Array.isArray(documentId) ? documentId : [documentId]).flatMap((id) => [
        getPublishedId$1(id),
        getDraftId(id)
      ]) : Array.isArray(documentId) ? documentId : [documentId]
    )
  ), subscriptionId = insecureRandomId();
  return state.set(
    "addSubscribers",
    (prev) => documentIds.reduce(
      (acc, id) => addSubscriptionIdToDocument(acc, id, subscriptionId),
      prev
    )
  ), () => {
    setTimeout(() => {
      state.set(
        "removeSubscribers",
        (prev) => documentIds.reduce(
          (acc, id) => removeSubscriptionIdFromDocument(acc, id, subscriptionId),
          prev
        )
      );
    }, DOCUMENT_STATE_CLEAR_DELAY);
  };
}
function getDocumentIdsFromActions(actions) {
  return Array.from(
    new Set(
      actions.map((i) => i.documentId).filter((i) => typeof i == "string").flatMap((documentId) => [getPublishedId$1(documentId), getDraftId(documentId)])
    )
  );
}
function getDocumentEvents(outgoing) {
  const documentIdsByAction = Object.entries(
    outgoing.actions.reduce(
      (acc, { type, documentId }) => {
        const ids = acc[type] || /* @__PURE__ */ new Set();
        return documentId && ids.add(documentId), acc[type] = ids, acc;
      },
      {}
    )
  ), actionMap = {
    "document.create": "created",
    "document.delete": "deleted",
    "document.discard": "discarded",
    "document.edit": "edited",
    "document.publish": "published",
    "document.unpublish": "unpublished"
  };
  return documentIdsByAction.flatMap(
    ([actionType, documentIds]) => Array.from(documentIds).map(
      (documentId) => ({ type: actionMap[actionType], documentId, outgoing })
    )
  );
}
const API_VERSION$2 = "v2025-05-06";
function createSharedListener(instance) {
  const dispose$ = new Subject(), events$ = getClientState(instance, {
    apiVersion: API_VERSION$2
  }).observable.pipe(
    switchMap(
      (client) => (
        // TODO: it seems like the client.listen method is not emitting disconnected
        // events. this is important to ensure we have an up to date version of the
        // doc. probably should introduce our own events for when the user goes offline
        client.listen(
          "*",
          {},
          {
            events: ["mutation", "welcome", "reconnect"],
            includeResult: !1,
            tag: "document-listener"
            // // from manual testing, it seems like mendoza patches may be
            // // causing some ambiguity/wonkiness
            // includeMutations: false,
            // effectFormat: 'mendoza',
          }
        )
      )
    ),
    takeUntil(dispose$),
    share()
  ), [welcome$, mutation$] = partition(events$, (e) => e.type === "welcome");
  return {
    events: merge(
      // we replay the welcome event because that event kicks off fetching the document
      welcome$.pipe(shareReplay(1)),
      mutation$
    ),
    dispose: () => dispose$.next()
  };
}
function createFetchDocument(instance) {
  return function(documentId) {
    return getClientState(instance, { apiVersion: API_VERSION$2 }).observable.pipe(
      switchMap((client) => createDocumentLoaderFromClient(client)(documentId)),
      map((result) => {
        if (!result.accessible) {
          if (result.reason === "existence") return null;
          throw new Error(`Document with ID \`${documentId}\` is inaccessible due to permissions.`);
        }
        return result.document;
      }),
      first$1()
    );
  };
}
const documentStore = {
  name: "Document",
  getInitialState: (instance) => ({
    documentStates: {},
    // these can be emptied on refetch
    queued: [],
    applied: [],
    sharedListener: createSharedListener(instance),
    fetchDocument: createFetchDocument(instance),
    events: new Subject()
  }),
  initialize(context) {
    const { sharedListener } = context.state.get(), subscriptions = [
      subscribeToQueuedAndApplyNextTransaction(context),
      subscribeToSubscriptionsAndListenToDocuments(context),
      subscribeToAppliedAndSubmitNextTransaction(context),
      subscribeToClientAndFetchDatasetAcl(context)
    ];
    return () => {
      sharedListener.dispose(), subscriptions.forEach((subscription) => subscription.unsubscribe());
    };
  }
};
function getDocumentState(...args) {
  return _getDocumentState(...args);
}
const _getDocumentState = bindActionByDataset(
  documentStore,
  createStateSourceAction({
    selector: ({ state: { error, documentStates } }, options) => {
      const { documentId, path, liveEdit } = options;
      if (error) throw error;
      if (liveEdit) {
        const document22 = documentStates[documentId]?.local;
        if (document22 === void 0) return;
        if (!path) return document22;
        const result2 = jsonMatch(document22, path).next();
        if (result2.done) return;
        const { value: value2 } = result2.value;
        return value2;
      }
      const draftId = getDraftId(documentId), publishedId = getPublishedId$1(documentId), draft = documentStates[draftId]?.local, published = documentStates[publishedId]?.local;
      if (draft === void 0 || published === void 0) return;
      const document2 = draft ?? published;
      if (!path) return document2;
      const result = jsonMatch(document2, path).next();
      if (result.done) return;
      const { value } = result.value;
      return value;
    },
    onSubscribe: (context, options) => manageSubscriberIds(context, options.documentId, { expandDraftPublished: !options.liveEdit })
  })
);
function resolveDocument(...args) {
  return _resolveDocument(...args);
}
const _resolveDocument = bindActionByDataset(
  documentStore,
  ({ instance }, docHandle) => firstValueFrom(
    getDocumentState(instance, {
      ...docHandle,
      path: void 0
    }).observable.pipe(filter((i) => i !== void 0))
  )
), getDocumentSyncStatus = bindActionByDataset(
  documentStore,
  createStateSourceAction({
    selector: ({ state: { error, documentStates: documents, outgoing, applied, queued } }, doc) => {
      const documentId = typeof doc == "string" ? doc : doc.documentId;
      if (error) throw error;
      if (doc.liveEdit)
        return documents[documentId] === void 0 ? void 0 : !queued.length && !applied.length && !outgoing;
      const draftId = getDraftId(documentId), publishedId = getPublishedId$1(documentId), draft = documents[draftId], published = documents[publishedId];
      if (!(draft === void 0 || published === void 0))
        return !queued.length && !applied.length && !outgoing;
    },
    onSubscribe: (context, doc) => manageSubscriberIds(context, doc.documentId)
  })
), getPermissionsState = bindActionByDataset(
  documentStore,
  createStateSourceAction({
    selector: calculatePermissions,
    onSubscribe: (context, { actions }) => manageSubscriberIds(context, getDocumentIdsFromActions(actions))
  })
), resolvePermissions = bindActionByDataset(
  documentStore,
  ({ instance }, options) => firstValueFrom(
    getPermissionsState(instance, options).observable.pipe(filter((i) => i !== void 0))
  )
), subscribeDocumentEvents = bindActionByDataset(
  documentStore,
  ({ state }, eventHandler) => {
    const { events } = state.get(), subscription = events.subscribe(eventHandler);
    return () => subscription.unsubscribe();
  }
), subscribeToQueuedAndApplyNextTransaction = ({ state }) => {
  const { events } = state.get();
  return state.observable.pipe(
    map(applyFirstQueuedTransaction),
    distinctUntilChanged(),
    tap$1((next) => state.set("applyFirstQueuedTransaction", next)),
    catchError$1((error, caught) => {
      if (error instanceof ActionError)
        return state.set(
          "removeQueuedTransaction",
          (prev) => removeQueuedTransaction(prev, error.transactionId)
        ), events.next({
          type: "error",
          message: error.message,
          documentId: error.documentId,
          transactionId: error.transactionId,
          error
        }), caught;
      throw error;
    })
  ).subscribe({ error: (error) => state.set("setError", { error }) });
}, subscribeToAppliedAndSubmitNextTransaction = ({
  state,
  instance
}) => {
  const { events } = state.get();
  return state.observable.pipe(
    throttle(
      (s) => (
        // if there is no outgoing transaction, we can throttle by the
        // initial outgoing throttle time…
        s.outgoing ? (
          // …otherwise, wait until the outgoing has been cleared
          state.observable.pipe(first$1(({ outgoing }) => !outgoing))
        ) : timer(INITIAL_OUTGOING_THROTTLE_TIME)
      ),
      { leading: !1, trailing: !0 }
    ),
    map(transitionAppliedTransactionsToOutgoing),
    distinctUntilChanged((a, b) => a.outgoing?.transactionId === b.outgoing?.transactionId),
    tap$1((next) => state.set("transitionAppliedTransactionsToOutgoing", next)),
    map((s) => s.outgoing),
    distinctUntilChanged(),
    withLatestFrom(getClientState(instance, { apiVersion: API_VERSION$3 }).observable),
    concatMap(([outgoing, client]) => outgoing ? client.observable.action(outgoing.outgoingActions, {
      transactionId: outgoing.transactionId,
      skipCrossDatasetReferenceValidation: !0
    }).pipe(
      catchError$1((error) => (state.set("revertOutgoingTransaction", revertOutgoingTransaction), events.next({ type: "reverted", message: error.message, outgoing, error }), EMPTY)),
      map((result) => ({ result, outgoing }))
    ) : EMPTY),
    tap$1(({ outgoing, result }) => {
      state.set("cleanupOutgoingTransaction", cleanupOutgoingTransaction);
      for (const e of getDocumentEvents(outgoing)) events.next(e);
      events.next({ type: "accepted", outgoing, result });
    })
  ).subscribe({ error: (error) => state.set("setError", { error }) });
}, subscribeToSubscriptionsAndListenToDocuments = (context) => {
  const { state } = context, { events } = state.get();
  return state.observable.pipe(
    filter((s) => !!s.grants),
    map((s) => Object.keys(s.documentStates)),
    distinctUntilChanged((curr, next) => {
      if (curr.length !== next.length) return !1;
      const currSet = new Set(curr);
      return next.every((i) => currSet.has(i));
    }),
    startWith$1(/* @__PURE__ */ new Set()),
    pairwise$1(),
    switchMap((pair) => {
      const [curr, next] = pair.map((ids) => new Set(ids)), added = Array.from(next).filter((i) => !curr.has(i)), removed = Array.from(curr).filter((i) => !next.has(i)), changes = [
        ...added.map((id) => ({ id, add: !0 })),
        ...removed.map((id) => ({ id, add: !1 }))
      ].sort((a, b) => {
        const aIsDraft = a.id === getDraftId(a.id), bIsDraft = b.id === getDraftId(b.id);
        return aIsDraft && bIsDraft ? a.id.localeCompare(b.id, "en-US") : aIsDraft ? -1 : bIsDraft ? 1 : a.id.localeCompare(b.id, "en-US");
      });
      return of(...changes);
    }),
    groupBy$1((i) => i.id),
    mergeMap$1(
      (group) => group.pipe(
        switchMap((e) => e.add ? listen(context, e.id).pipe(
          catchError$1((error) => {
            throw error instanceof OutOfSyncError && listen(context, e.id), error;
          }),
          tap$1(
            (remote) => state.set(
              "applyRemoteDocument",
              (prev) => applyRemoteDocument(prev, remote, events)
            )
          )
        ) : EMPTY)
      )
    )
  ).subscribe({ error: (error) => state.set("setError", { error }) });
}, subscribeToClientAndFetchDatasetAcl = ({
  instance,
  state,
  key: { projectId, dataset }
}) => getClientState(instance, { apiVersion: API_VERSION$3 }).observable.pipe(
  switchMap(
    (client) => client.observable.request({
      uri: `/projects/${projectId}/datasets/${dataset}/acl`,
      tag: "acl.get",
      withCredentials: !0
    })
  ),
  tap$1((datasetAcl) => state.set("setGrants", { grants: createGrantsLookup(datasetAcl) }))
).subscribe({
  error: (error) => state.set("setError", { error })
});
function applyDocumentActions(...args) {
  return boundApplyDocumentActions(...args);
}
const boundApplyDocumentActions = bindActionByDataset(documentStore, _applyDocumentActions);
async function _applyDocumentActions({ state }, { actions, transactionId = crypto.randomUUID(), disableBatching }) {
  const { events } = state.get(), transaction = {
    transactionId,
    actions,
    ...disableBatching && { disableBatching }
  }, fatalError$ = state.observable.pipe(
    map((s) => s.error),
    first$1(Boolean),
    map((error) => ({ type: "error", error }))
  ), transactionError$ = events.pipe(
    filter((e) => e.type === "error"),
    first$1((e) => e.transactionId === transactionId)
  ), appliedTransaction$ = state.observable.pipe(
    map((s) => s.applied),
    distinctUntilChanged(),
    map((applied) => applied.find((t) => t.transactionId === transactionId)),
    first$1(Boolean)
  ), successfulTransaction$ = events.pipe(
    filter((e) => e.type === "accepted"),
    first$1((e) => e.outgoing.batchedTransactionIds.includes(transactionId))
  ), rejectedTransaction$ = events.pipe(
    filter((e) => e.type === "reverted"),
    first$1((e) => e.outgoing.batchedTransactionIds.includes(transactionId))
  ), appliedTransactionOrError = firstValueFrom(
    race([fatalError$, transactionError$, appliedTransaction$])
  ), acceptedOrRejectedTransaction = firstValueFrom(
    race([successfulTransaction$, rejectedTransaction$, transactionError$])
  );
  state.set("queueTransaction", (prev) => queueTransaction(prev, transaction));
  const result = await appliedTransactionOrError;
  if ("type" in result && result.type === "error") throw result.error;
  const { working: documents, previous, previousRevs } = result, existingIds = new Set(
    Object.entries(previous).filter(([, value]) => !!value).map(([key]) => key)
  ), resultingIds = new Set(
    Object.entries(documents).filter(([, value]) => !!value).map(([key]) => key)
  ), allIds = /* @__PURE__ */ new Set([...existingIds, ...resultingIds]), updated = [], appeared = [], disappeared = [];
  for (const id of allIds)
    existingIds.has(id) && resultingIds.has(id) ? updated.push(id) : !existingIds.has(id) && resultingIds.has(id) ? appeared.push(id) : !resultingIds.has(id) && existingIds.has(id) && disappeared.push(id);
  async function submitted() {
    const raceResult = await acceptedOrRejectedTransaction;
    if (raceResult.type !== "accepted") throw raceResult.error;
    return raceResult.result;
  }
  return {
    transactionId,
    documents,
    previous,
    previousRevs,
    appeared,
    updated,
    disappeared,
    submitted
  };
}
function createFavoriteKey(context) {
  return `${context.documentId}:${context.documentType}:${context.resourceId}:${context.resourceType}${context.schemaName ? `:${context.schemaName}` : ""}`;
}
const favorites = createFetcherStore({
  name: "Favorites",
  getKey: (_instance, context) => createFavoriteKey(context),
  fetcher: (instance) => (context) => {
    const nodeStateSource = getNodeState(instance, {
      name: SDK_NODE_NAME,
      connectTo: SDK_CHANNEL_NAME
    }), payload = {
      document: {
        id: context.documentId,
        type: context.documentType,
        resource: {
          id: context.resourceId,
          type: context.resourceType,
          schemaName: context.schemaName
        }
      }
    };
    return nodeStateSource.observable.pipe(
      filter((nodeState) => !!nodeState),
      // Only proceed when connected
      shareReplay(1),
      switchMap((nodeState) => {
        const node = nodeState.node;
        return from(
          node.fetch(
            // @ts-expect-error -- getOrCreateNode should be refactored to take type arguments
            "dashboard/v1/events/favorite/query",
            payload
          )
        ).pipe(
          map((response) => ({ isFavorited: response.isFavorited })),
          catchError$1((err) => (console.error("Favorites service connection error", err), of({ isFavorited: !1 })))
        );
      })
    );
  }
}), getFavoritesState = favorites.getState, resolveFavoritesState = favorites.resolveState, API_VERSION$1 = "vX", PROJECT_API_VERSION = "2025-07-18", USERS_STATE_CLEAR_DELAY = 5e3, DEFAULT_USERS_BATCH_SIZE = 100, getUsersKey = (instance, {
  resourceType,
  organizationId,
  batchSize = DEFAULT_USERS_BATCH_SIZE,
  projectId = instance.config.projectId,
  userId
} = {}) => JSON.stringify({
  resourceType,
  organizationId,
  batchSize,
  projectId,
  userId
}), parseUsersKey = (key) => JSON.parse(key), addSubscription = (subscriptionId, key) => (prev) => {
  const group = prev.users[key], subscriptions = [...group?.subscriptions ?? [], subscriptionId];
  return { ...prev, users: { ...prev.users, [key]: { ...group, subscriptions } } };
}, removeSubscription = (subscriptionId, key) => (prev) => {
  const group = prev.users[key];
  if (!group) return prev;
  const subscriptions = group.subscriptions.filter((id) => id !== subscriptionId);
  return subscriptions.length ? { ...prev, users: { ...prev.users, [key]: { ...group, subscriptions } } } : { ...prev, users: omit(prev.users, key) };
}, setUsersData = (key, { data, nextCursor, totalCount }) => (prev) => {
  const group = prev.users[key];
  if (!group) return prev;
  const users = [...group.users ?? [], ...data];
  return { ...prev, users: { ...prev.users, [key]: { ...group, users, totalCount, nextCursor } } };
}, updateLastLoadMoreRequest = (timestamp, key) => (prev) => {
  const group = prev.users[key];
  return group ? { ...prev, users: { ...prev.users, [key]: { ...group, lastLoadMoreRequest: timestamp } } } : prev;
}, setUsersError = (key, error) => (prev) => {
  const group = prev.users[key];
  return group ? { ...prev, users: { ...prev.users, [key]: { ...group, error } } } : prev;
}, cancelRequest = (key) => (prev) => {
  const group = prev.users[key];
  return !group || group.subscriptions.length ? prev : { ...prev, users: omit(prev.users, key) };
}, initializeRequest = (key) => (prev) => prev.users[key] ? prev : { ...prev, users: { ...prev.users, [key]: { subscriptions: [] } } }, usersStore = {
  name: "UsersStore",
  getInitialState: () => ({ users: {} }),
  initialize: (context) => {
    const subscription = listenForLoadMoreAndFetch(context);
    return () => subscription.unsubscribe();
  }
}, errorHandler$1 = (state) => (error) => state.set("setError", { error }), listenForLoadMoreAndFetch = ({ state, instance }) => state.observable.pipe(
  map((s) => new Set(Object.keys(s.users))),
  distinctUntilChanged((curr, next) => curr.size !== next.size ? !1 : Array.from(next).every((i) => curr.has(i))),
  startWith$1(/* @__PURE__ */ new Set()),
  pairwise$1(),
  mergeMap$1(([curr, next]) => {
    const added = Array.from(next).filter((i) => !curr.has(i)), removed = Array.from(curr).filter((i) => !next.has(i));
    return [
      ...added.map((key) => ({ key, added: !0 })),
      ...removed.map((key) => ({ key, added: !1 }))
    ];
  }),
  groupBy$1((i) => i.key),
  mergeMap$1(
    (group$) => group$.pipe(
      switchMap((e) => {
        if (!e.added) return EMPTY;
        const { userId, batchSize, ...options } = parseUsersKey(group$.key);
        if (userId) {
          if (userId.startsWith("p"))
            return getClient(instance, {
              apiVersion: PROJECT_API_VERSION,
              // this is a global store, so we need to use the projectId from the options when we're fetching
              // users from a project subdomain
              projectId: options.projectId,
              useProjectHostname: !0
            }).observable.request({
              method: "GET",
              uri: `/users/${userId}`
            }).pipe(
              map((user) => ({
                data: [{
                  sanityUserId: user.sanityUserId,
                  profile: {
                    id: user.id,
                    displayName: user.displayName,
                    familyName: user.familyName ?? void 0,
                    givenName: user.givenName ?? void 0,
                    middleName: user.middleName ?? void 0,
                    imageUrl: user.imageUrl ?? void 0,
                    createdAt: user.createdAt,
                    updatedAt: user.updatedAt,
                    isCurrentUser: user.isCurrentUser,
                    email: user.email,
                    provider: user.provider
                  },
                  memberships: []
                }],
                totalCount: 1,
                nextCursor: null
              })),
              catchError$1((error) => (state.set("setUsersError", setUsersError(group$.key, error)), EMPTY)),
              tap$1(
                (response) => state.set("setUsersData", setUsersData(group$.key, response))
              )
            );
          const scope = userId.startsWith("g") ? "global" : void 0, client = getClient(instance, {
            scope,
            apiVersion: API_VERSION$1
          }), resourceType2 = options.resourceType || "project", resourceId = resourceType2 === "organization" ? options.organizationId : options.projectId;
          return resourceId ? client.observable.request({
            method: "GET",
            uri: `access/${resourceType2}/${resourceId}/users/${userId}`
          }).pipe(
            map((response) => "sanityUserId" in response ? {
              data: [response],
              totalCount: 1,
              nextCursor: null
            } : response),
            catchError$1((error) => (state.set("setUsersError", setUsersError(group$.key, error)), EMPTY)),
            tap$1((response) => state.set("setUsersData", setUsersData(group$.key, response)))
          ) : throwError(() => new Error("An organizationId or a projectId is required"));
        }
        const projectId = options.projectId, resourceType = options.resourceType ?? (options.organizationId ? "organization" : projectId ? "project" : "organization"), organizationId$ = options.organizationId ? of(options.organizationId) : getDashboardOrganizationId$1(instance).observable.pipe(
          filter((i) => typeof i == "string")
        ), resource$ = resourceType === "project" ? projectId ? of({ type: "project", id: projectId }) : throwError(() => new Error("Project ID required for this API.")) : organizationId$.pipe(map((id) => ({ type: "organization", id }))), client$ = getClientState(instance, {
          scope: "global",
          apiVersion: API_VERSION$1
        }).observable, loadMore$ = state.observable.pipe(
          map((s) => s.users[group$.key]?.lastLoadMoreRequest),
          distinctUntilChanged()
        ), cursor$ = state.observable.pipe(
          map((s) => s.users[group$.key]?.nextCursor),
          distinctUntilChanged(),
          filter((cursor) => cursor !== null)
        );
        return combineLatest([resource$, client$, loadMore$]).pipe(
          withLatestFrom(cursor$),
          switchMap(
            ([[resource, client], cursor]) => client.observable.request({
              method: "GET",
              uri: `access/${resource.type}/${resource.id}/users`,
              query: cursor ? { nextCursor: cursor, limit: batchSize.toString() } : { limit: batchSize.toString() }
            })
          ),
          catchError$1((error) => (state.set("setUsersError", setUsersError(group$.key, error)), EMPTY)),
          tap$1((response) => state.set("setUsersData", setUsersData(group$.key, response)))
        );
      })
    )
  )
).subscribe({ error: errorHandler$1(state) }), getUsersState = bindActionGlobally(
  usersStore,
  createStateSourceAction({
    selector: createSelector(
      [
        ({ instance, state }, options) => state.error ?? state.users[getUsersKey(instance, options)]?.error,
        ({ instance, state }, options) => state.users[getUsersKey(instance, options)]?.users,
        ({ instance, state }, options) => state.users[getUsersKey(instance, options)]?.totalCount,
        ({ instance, state }, options) => state.users[getUsersKey(instance, options)]?.nextCursor
      ],
      (error, data, totalCount, nextCursor) => {
        if (error) throw error;
        if (!(data === void 0 || totalCount === void 0 || nextCursor === void 0))
          return { data, totalCount, hasMore: nextCursor !== null };
      }
    ),
    onSubscribe: ({ instance, state }, options) => {
      const subscriptionId = insecureRandomId(), key = getUsersKey(instance, options);
      return state.set("addSubscription", addSubscription(subscriptionId, key)), () => {
        setTimeout(
          () => state.set("removeSubscription", removeSubscription(subscriptionId, key)),
          USERS_STATE_CLEAR_DELAY
        );
      };
    }
  })
), resolveUsers = bindActionGlobally(
  usersStore,
  async ({ state, instance }, { signal, ...options }) => {
    const key = getUsersKey(instance, options), { getCurrent } = getUsersState(instance, options), aborted$ = signal ? new Observable((observer) => {
      const cleanup = () => {
        signal.removeEventListener("abort", listener);
      }, listener = () => {
        observer.error(new DOMException("The operation was aborted.", "AbortError")), observer.complete(), cleanup();
      };
      return signal.addEventListener("abort", listener), cleanup;
    }).pipe(
      catchError$1((error) => {
        throw error instanceof Error && error.name === "AbortError" && state.set("cancelRequest", cancelRequest(key)), error;
      })
    ) : NEVER;
    state.set("initializeRequest", initializeRequest(key));
    const resolved$ = state.observable.pipe(
      map(getCurrent),
      first$1((i) => i !== void 0)
    );
    return firstValueFrom(race([resolved$, aborted$]));
  }
), loadMoreUsers = bindActionGlobally(
  usersStore,
  async ({ state, instance }, options) => {
    const key = getUsersKey(instance, options), users = getUsersState(instance, options), usersState = users.getCurrent();
    if (!usersState)
      throw new Error("Users not loaded for specified resource. Please call resolveUsers first.");
    if (!usersState.hasMore)
      throw new Error("No more users available to load for this resource.");
    const promise = firstValueFrom(
      users.observable.pipe(
        filter((i) => i !== void 0),
        skip(1)
      )
    ), timestamp = (/* @__PURE__ */ new Date()).toISOString();
    return state.set("updateLastLoadMoreRequest", updateLastLoadMoreRequest(timestamp, key)), await promise;
  }
), getUserState = bindActionGlobally(
  usersStore,
  ({ instance }, { userId, ...options }) => getUsersState(instance, { userId, ...options }).observable.pipe(
    map((res) => res?.data[0]),
    distinctUntilChanged((a, b) => a?.profile.updatedAt === b?.profile.updatedAt)
  )
), resolveUser = bindActionGlobally(
  usersStore,
  async ({ instance }, { signal, ...options }) => (await resolveUsers(instance, {
    signal,
    ...options
  }))?.data[0]
);
function getBifurClient(client, token$) {
  const bifurVersionedClient = client.withConfig({ apiVersion: "2022-06-30" }), { dataset, url: baseUrl, requestTagPrefix = "sanity.studio" } = bifurVersionedClient.config(), urlWithTag = `${`${baseUrl.replace(/\/+$/, "")}/socket/${dataset}`.replace(/^http/, "ws")}?tag=${requestTagPrefix}`;
  return fromUrl(urlWithTag, { token$ });
}
const handleIncomingMessage = (event) => {
  switch (event.type) {
    case "rollCall":
      return {
        type: "rollCall",
        userId: event.i,
        sessionId: event.session
      };
    case "state": {
      const { sessionId, locations } = event.m;
      return {
        type: "state",
        userId: event.i,
        sessionId,
        timestamp: (/* @__PURE__ */ new Date()).toISOString(),
        locations
      };
    }
    case "disconnect":
      return {
        type: "disconnect",
        userId: event.i,
        sessionId: event.m.session,
        timestamp: (/* @__PURE__ */ new Date()).toISOString()
      };
    default:
      throw new Error(`Got unknown presence event: ${JSON.stringify(event)}`);
  }
}, createBifurTransport = (options) => {
  const { client, token$, sessionId } = options, bifur = getBifurClient(client, token$), incomingEvents$ = bifur.listen("presence").pipe(map$1(handleIncomingMessage)), dispatchMessage = (message) => {
    switch (message.type) {
      case "rollCall":
        return bifur.request("presence_rollcall", { session: sessionId });
      case "state":
        return bifur.request("presence_announce", {
          data: { locations: message.locations, sessionId }
        });
      case "disconnect":
        return bifur.request("presence_disconnect", { session: sessionId });
      default:
        return EMPTY;
    }
  };
  return typeof window < "u" && fromEvent(window, "beforeunload").pipe(switchMap$1(() => dispatchMessage({ type: "disconnect" }))).subscribe(), [incomingEvents$.pipe(share$1()), dispatchMessage];
}, getInitialState = () => ({
  locations: /* @__PURE__ */ new Map(),
  users: {}
}), presenceStore = {
  name: "presence",
  getInitialState,
  initialize: (context) => {
    const {
      instance,
      state,
      key: { projectId, dataset }
    } = context, sessionId = crypto.randomUUID(), client = getClient(instance, {
      apiVersion: "2022-06-30",
      projectId,
      dataset
    }), token$ = getTokenState(instance).observable.pipe(distinctUntilChanged()), [incomingEvents$, dispatch] = createBifurTransport({
      client,
      token$,
      sessionId
    }), subscription = new Subscription();
    return subscription.add(
      incomingEvents$.subscribe((event) => {
        "sessionId" in event && event.sessionId === sessionId || (event.type === "state" ? state.set("presence/state", (prevState) => {
          const newLocations = new Map(prevState.locations);
          return newLocations.set(event.sessionId, {
            userId: event.userId,
            locations: event.locations
          }), {
            ...prevState,
            locations: newLocations
          };
        }) : event.type === "disconnect" && state.set("presence/disconnect", (prevState) => {
          const newLocations = new Map(prevState.locations);
          return newLocations.delete(event.sessionId), { ...prevState, locations: newLocations };
        }));
      })
    ), dispatch({ type: "rollCall" }).subscribe(), () => {
      dispatch({ type: "disconnect" }).subscribe(), subscription.unsubscribe();
    };
  }
}, selectLocations = (state) => state.locations, selectUsers = (state) => state.users, selectPresence = createSelector(
  selectLocations,
  selectUsers,
  (locations, users) => Array.from(locations.entries()).map(([sessionId, { userId, locations: locs }]) => ({
    user: users[userId] || {
      id: userId,
      displayName: "Unknown user",
      name: "Unknown user",
      email: ""
    },
    sessionId,
    locations: locs
  }))
), getPresence = bindActionByDataset(
  presenceStore,
  createStateSourceAction({
    selector: (context, _) => selectPresence(context.state),
    onSubscribe: (context, _) => {
      const subscription = context.state.observable.pipe(
        map(
          (state) => Array.from(state.locations.values()).map((l) => l.userId).filter((id) => !!id)
        ),
        distinctUntilChanged((a, b) => a.length === b.length && a.every((v, i) => v === b[i]))
      ).pipe(
        switchMap((userIds) => {
          if (userIds.length === 0)
            return of([]);
          const userObservables = userIds.map(
            (userId) => getUserState(context.instance, {
              userId,
              resourceType: "project",
              projectId: context.key.projectId
            }).pipe(filter((v) => !!v))
          );
          return combineLatest(userObservables);
        })
      ).subscribe((users) => {
        users && context.state.set("presence/users", (prevState) => ({
          ...prevState,
          users: {
            ...prevState.users,
            ...users.reduce((acc, user) => (user && (acc[user.profile.id] = user), acc), {})
          }
        }));
      });
      return () => subscription.unsubscribe();
    }
  })
);
function sortReleases(releases = []) {
  return [...releases].sort((a, b) => {
    if (a.metadata.releaseType === "undecided" && b.metadata.releaseType !== "undecided")
      return -1;
    if (a.metadata.releaseType !== "undecided" && b.metadata.releaseType === "undecided")
      return 1;
    if (a.metadata.releaseType === "undecided" && b.metadata.releaseType === "undecided")
      return new Date(b._createdAt).getTime() - new Date(a._createdAt).getTime();
    if (a.metadata.releaseType === "scheduled" && b.metadata.releaseType === "scheduled") {
      const aPublishAt = a.publishAt || a.metadata.intendedPublishAt;
      if (!aPublishAt)
        return 1;
      const bPublishAt = b.publishAt || b.metadata.intendedPublishAt;
      return bPublishAt ? new Date(bPublishAt).getTime() - new Date(aPublishAt).getTime() : -1;
    }
    return a.metadata.releaseType === "asap" && b.metadata.releaseType !== "asap" ? 1 : a.metadata.releaseType !== "asap" && b.metadata.releaseType === "asap" ? -1 : a.metadata.releaseType === "asap" && b.metadata.releaseType === "asap" ? new Date(b._createdAt).getTime() - new Date(a._createdAt).getTime() : 0;
  });
}
const ARCHIVED_RELEASE_STATES = ["archived", "published"], STABLE_EMPTY_RELEASES = [], releasesStore = {
  name: "Releases",
  getInitialState: () => ({
    activeReleases: void 0
  }),
  initialize: (context) => {
    const subscription = subscribeToReleases(context);
    return () => subscription.unsubscribe();
  }
}, getActiveReleasesState = bindActionByDataset(
  releasesStore,
  createStateSourceAction({
    selector: ({ state }, _) => state.activeReleases
  })
), RELEASES_QUERY = "releases::all()", subscribeToReleases = ({
  instance,
  state,
  key: { projectId, dataset }
}) => {
  const { observable: releases$ } = getQueryState(instance, {
    query: RELEASES_QUERY,
    perspective: "raw",
    projectId,
    dataset,
    tag: "releases"
  });
  return releases$.pipe(
    map((releases) => {
      state.set("setActiveReleases", {
        activeReleases: sortReleases(releases ?? STABLE_EMPTY_RELEASES).filter((release) => !ARCHIVED_RELEASE_STATES.includes(release.state)).reverse()
      });
    })
  ).subscribe({ error: (error) => state.set("setError", { error }) });
}, DEFAULT_PERSPECTIVE = "drafts", optionsCache = /* @__PURE__ */ new Map(), selectInstancePerspective = (context, _) => context.instance.config.perspective, selectActiveReleases = (context) => context.state.activeReleases, selectOptions = (_context, options) => options, memoizedOptionsSelector = createSelector(
  [selectActiveReleases, selectOptions],
  (activeReleases, options) => {
    if (!options || !activeReleases) return options;
    const releaseIds = activeReleases.map((release) => release._id).join(",");
    let nestedCache = optionsCache.get(releaseIds);
    nestedCache || (nestedCache = /* @__PURE__ */ new Map(), optionsCache.set(releaseIds, nestedCache));
    const optionsKey = JSON.stringify(options);
    let cachedOptions = nestedCache.get(optionsKey);
    return cachedOptions || (cachedOptions = options, nestedCache.set(optionsKey, cachedOptions)), cachedOptions;
  }
), _getPerspectiveStateSelector = createStateSourceAction({
  selector: createSelector(
    [selectInstancePerspective, selectActiveReleases, memoizedOptionsSelector],
    (instancePerspective, activeReleases, memoizedOptions) => {
      const perspective = memoizedOptions?.perspective ?? instancePerspective ?? DEFAULT_PERSPECTIVE;
      if (!isReleasePerspective(perspective)) return perspective;
      if (!activeReleases || activeReleases.length === 0) return;
      const releaseNames = sortReleases(activeReleases).map((release) => release.name), index = releaseNames.findIndex((name) => name === perspective.releaseName);
      if (index < 0)
        throw new Error(`Release "${perspective.releaseName}" not found in active releases`);
      return ["drafts", ...releaseNames.slice(0, index + 1)].filter((name) => !perspective.excludedPerspectives?.includes(name)).reverse();
    }
  )
});
let _boundGetPerspectiveState;
const getPerspectiveState = (...args) => (_boundGetPerspectiveState || (_boundGetPerspectiveState = bindActionByDataset(
  releasesStore,
  _getPerspectiveStateSelector
)), _boundGetPerspectiveState(...args)), QUERY_STATE_CLEAR_DELAY = 1e3, QUERY_STORE_API_VERSION = "v2025-05-06", QUERY_STORE_DEFAULT_PERSPECTIVE = "drafts", setQueryError = (key, error) => (prev) => {
  const prevQuery = prev.queries[key];
  return prevQuery ? { ...prev, queries: { ...prev.queries, [key]: { ...prevQuery, error } } } : prev;
}, setQueryData = (key, result, syncTags) => (prev) => {
  const prevQuery = prev.queries[key];
  return prevQuery ? {
    ...prev,
    queries: { ...prev.queries, [key]: { ...prevQuery, result: result ?? null, syncTags } }
  } : prev;
}, setLastLiveEventId = (key, lastLiveEventId) => (prev) => {
  const prevQuery = prev.queries[key];
  return prevQuery ? { ...prev, queries: { ...prev.queries, [key]: { ...prevQuery, lastLiveEventId } } } : prev;
}, addSubscriber = (key, subscriptionId) => (prev) => {
  const prevQuery = prev.queries[key], subscribers = [...prevQuery?.subscribers ?? [], subscriptionId];
  return { ...prev, queries: { ...prev.queries, [key]: { ...prevQuery, subscribers } } };
}, removeSubscriber = (key, subscriptionId) => (prev) => {
  const prevQuery = prev.queries[key];
  if (!prevQuery) return prev;
  const subscribers = prevQuery.subscribers.filter((id) => id !== subscriptionId);
  return subscribers.length ? { ...prev, queries: { ...prev.queries, [key]: { ...prevQuery, subscribers } } } : { ...prev, queries: omit(prev.queries, key) };
}, cancelQuery = (key) => (prev) => {
  const prevQuery = prev.queries[key];
  return !prevQuery || prevQuery.subscribers.length ? prev : { ...prev, queries: omit(prev.queries, key) };
}, initializeQuery = (key) => (prev) => prev.queries[key] ? prev : { ...prev, queries: { ...prev.queries, [key]: { subscribers: [] } } }, EMPTY_ARRAY = [], getQueryKey = (options) => JSON.stringify(options), parseQueryKey = (key) => JSON.parse(key);
function normalizeOptionsWithPerspective(instance, options) {
  if (options.perspective !== void 0) return options;
  const instancePerspective = instance.config.perspective;
  return {
    ...options,
    perspective: instancePerspective !== void 0 ? instancePerspective : QUERY_STORE_DEFAULT_PERSPECTIVE
  };
}
const queryStore = {
  name: "QueryStore",
  getInitialState: () => ({ queries: {} }),
  initialize(context) {
    const subscriptions = [
      listenForNewSubscribersAndFetch(context),
      listenToLiveClientAndSetLastLiveEventIds(context)
    ];
    return () => {
      for (const subscription of subscriptions)
        subscription.unsubscribe();
    };
  }
}, errorHandler = (state) => (error) => state.set("setError", { error }), listenForNewSubscribersAndFetch = ({ state, instance }) => state.observable.pipe(
  map((s) => new Set(Object.keys(s.queries))),
  distinctUntilChanged((curr, next) => curr.size !== next.size ? !1 : Array.from(next).every((i) => curr.has(i))),
  startWith$1(/* @__PURE__ */ new Set()),
  pairwise$1(),
  mergeMap$1(([curr, next]) => {
    const added = Array.from(next).filter((i) => !curr.has(i)), removed = Array.from(curr).filter((i) => !next.has(i));
    return [
      ...added.map((key) => ({ key, added: !0 })),
      ...removed.map((key) => ({ key, added: !1 }))
    ];
  }),
  groupBy$1((i) => i.key),
  mergeMap$1(
    (group$) => group$.pipe(
      switchMap((e) => {
        if (!e.added) return EMPTY;
        const lastLiveEventId$ = state.observable.pipe(
          map((s) => s.queries[group$.key]?.lastLiveEventId),
          distinctUntilChanged()
        ), {
          query,
          params,
          projectId,
          dataset,
          tag,
          source,
          perspective: perspectiveFromOptions,
          ...restOptions
        } = parseQueryKey(group$.key), perspective$ = isReleasePerspective(perspectiveFromOptions) ? getPerspectiveState(instance, {
          perspective: perspectiveFromOptions
        }).observable.pipe(filter(Boolean)) : of(perspectiveFromOptions ?? QUERY_STORE_DEFAULT_PERSPECTIVE), client$ = getClientState(instance, {
          apiVersion: QUERY_STORE_API_VERSION,
          projectId,
          dataset,
          source
        }).observable;
        return combineLatest({
          lastLiveEventId: lastLiveEventId$,
          client: client$,
          perspective: perspective$
        }).pipe(
          switchMap(
            ({ lastLiveEventId, client, perspective }) => client.observable.fetch(query, params, {
              ...restOptions,
              perspective,
              filterResponse: !1,
              returnQuery: !1,
              lastLiveEventId,
              tag
            })
          )
        );
      }),
      catchError$1((error) => (state.set("setQueryError", setQueryError(group$.key, error)), EMPTY)),
      tap$1(({ result, syncTags }) => {
        state.set("setQueryData", setQueryData(group$.key, result, syncTags));
      })
    )
  )
).subscribe({ error: errorHandler(state) }), listenToLiveClientAndSetLastLiveEventIds = ({
  state,
  instance,
  key: { source }
}) => {
  const liveMessages$ = getClientState(instance, {
    apiVersion: QUERY_STORE_API_VERSION,
    // temporary guard here until we're ready for everything to be queried via global api
    ...source && !isDatasetSource(source) ? { source } : {}
  }).observable.pipe(
    switchMap(
      (client) => defer(
        () => client.live.events({ includeDrafts: !!client.config().token, tag: "query-store" })
      ).pipe(
        catchError$1((error) => {
          if (error instanceof CorsOriginError)
            return state.set("setError", { error }), EMPTY;
          throw error;
        })
      )
    ),
    share(),
    filter((e) => e.type === "message")
  );
  return state.observable.pipe(
    mergeMap$1((s) => Object.entries(s.queries)),
    groupBy$1(([key]) => key),
    mergeMap$1((group$) => {
      const syncTags$ = group$.pipe(
        map(([, queryState]) => queryState),
        map((i) => i?.syncTags ?? EMPTY_ARRAY),
        distinctUntilChanged()
      );
      return combineLatest([liveMessages$, syncTags$]).pipe(
        filter(([message, syncTags]) => message.tags.some((tag) => syncTags.includes(tag))),
        tap$1(([message]) => {
          state.set("setLastLiveEventId", setLastLiveEventId(group$.key, message.id));
        })
      );
    })
  ).subscribe({ error: errorHandler(state) });
};
function getQueryState(...args) {
  return _getQueryState(...args);
}
const _getQueryState = bindActionBySource(
  queryStore,
  createStateSourceAction({
    selector: ({ state, instance }, options) => {
      if (state.error) throw state.error;
      const key = getQueryKey(normalizeOptionsWithPerspective(instance, options)), queryState = state.queries[key];
      if (queryState?.error) throw queryState.error;
      return queryState?.result;
    },
    onSubscribe: ({ state, instance }, options) => {
      const subscriptionId = insecureRandomId(), key = getQueryKey(normalizeOptionsWithPerspective(instance, options));
      return state.set("addSubscriber", addSubscriber(key, subscriptionId)), () => {
        setTimeout(
          () => state.set("removeSubscriber", removeSubscriber(key, subscriptionId)),
          QUERY_STATE_CLEAR_DELAY
        );
      };
    }
  })
);
function resolveQuery(...args) {
  return _resolveQuery(...args);
}
const _resolveQuery = bindActionBySource(
  queryStore,
  ({ state, instance }, { signal, ...options }) => {
    const normalized = normalizeOptionsWithPerspective(instance, options), { getCurrent } = getQueryState(instance, normalized), key = getQueryKey(normalized), aborted$ = signal ? new Observable((observer) => {
      const cleanup = () => {
        signal.removeEventListener("abort", listener);
      }, listener = () => {
        observer.error(new DOMException("The operation was aborted.", "AbortError")), observer.complete(), cleanup();
      };
      return signal.addEventListener("abort", listener), cleanup;
    }).pipe(
      catchError$1((error) => {
        throw error instanceof Error && error.name === "AbortError" && state.set("cancelQuery", cancelQuery(key)), error;
      })
    ) : NEVER;
    state.set("initializeQuery", initializeQuery(key));
    const resolved$ = state.observable.pipe(
      map(getCurrent),
      first$1((i) => i !== void 0)
    );
    return firstValueFrom(race([resolved$, aborted$]));
  }
);
function hashString(str) {
  let hash = 0;
  for (let i = 0; i < str.length; i++)
    hash = (hash * 31 + str.charCodeAt(i)) % 2147483647;
  return Math.abs(hash).toString(16).padStart(8, "0");
}
const TITLE_CANDIDATES = ["title", "name", "label", "heading", "header", "caption"], SUBTITLE_CANDIDATES = ["description", "subtitle", ...TITLE_CANDIDATES], PREVIEW_PROJECTION = `{
  // Get all potential title fields
  "titleCandidates": {
    ${TITLE_CANDIDATES.map((field) => `"${field}": ${field}`).join(`,
      `)}
  },
  // Get all potential subtitle fields
  "subtitleCandidates": {
    ${SUBTITLE_CANDIDATES.map((field) => `"${field}": ${field}`).join(`,
      `)}
  },
  "media": coalesce(
    select(
      defined(asset) => {"type": "image-asset", "_ref": asset._ref},
      defined(image.asset) => {"type": "image-asset", "_ref": image.asset._ref},
      defined(mainImage.asset) => {"type": "image-asset", "_ref": mainImage.asset._ref},
      null
    )
  ),
  _type,
  _id,
  _updatedAt
}`, PREVIEW_TAG = "preview", PREVIEW_PERSPECTIVE = "raw", STABLE_EMPTY_PREVIEW = { data: null, isPending: !1 }, STABLE_ERROR_PREVIEW = {
  data: {
    title: "Preview Error",
    ...!!getEnv("DEV") && { subtitle: "Check the console for more details" }
  },
  isPending: !1
};
function assetIdToUrl(assetId, projectId, dataset) {
  const pattern = /^image-(?<assetName>[A-Za-z0-9]+)-(?<dimensions>\d+x\d+)-(?<format>[a-z]+)$/, match = assetId.match(pattern);
  if (!match?.groups)
    throw new Error(
      `Invalid asset ID \`${assetId}\`. Expected: image-{assetName}-{width}x{height}-{format}`
    );
  const { assetName, dimensions, format } = match.groups;
  return `https://cdn.sanity.io/images/${projectId}/${dataset}/${assetName}-${dimensions}.${format}`;
}
function hasImageRef(value) {
  return isObject(value) && "_ref" in value && typeof value._ref == "string";
}
function normalizeMedia(media, projectId, dataset) {
  return !media || !hasImageRef(media) ? null : {
    type: "image-asset",
    _ref: media._ref,
    url: assetIdToUrl(media._ref, projectId, dataset)
  };
}
function findFirstDefined(fieldsToSearch, candidates, exclude) {
  if (candidates)
    for (const field of fieldsToSearch) {
      const value = candidates[field];
      if (typeof value == "string" && value.trim() !== "" && value !== exclude)
        return value;
    }
}
function processPreviewQuery({
  projectId,
  dataset,
  ids,
  results
}) {
  const resultMap = results.reduce((acc, next) => (acc[next._id] = next, acc), {});
  return Object.fromEntries(
    Array.from(ids).map((id) => {
      const publishedId = getPublishedId(id), draftId = getDraftId(id), draftResult = resultMap[draftId], publishedResult = resultMap[publishedId];
      if (!draftResult && !publishedResult) return [id, STABLE_EMPTY_PREVIEW];
      try {
        const result = draftResult || publishedResult;
        if (!result) return [id, STABLE_EMPTY_PREVIEW];
        const title = findFirstDefined(TITLE_CANDIDATES, result.titleCandidates), subtitle = findFirstDefined(SUBTITLE_CANDIDATES, result.subtitleCandidates, title), preview = {
          title: String(title || `${result._type}: ${result._id}`),
          subtitle: subtitle || void 0,
          media: normalizeMedia(result.media, projectId, dataset)
        }, _status = {
          ...draftResult?._updatedAt && { lastEditedDraftAt: draftResult._updatedAt },
          ...publishedResult?._updatedAt && { lastEditedPublishedAt: publishedResult._updatedAt }
        };
        return [id, { data: { ...preview, _status }, isPending: !1 }];
      } catch (e) {
        return console.warn(e), [id, STABLE_ERROR_PREVIEW];
      }
    })
  );
}
function createPreviewQuery(documentIds) {
  const allIds = Array.from(documentIds).flatMap((id) => [getPublishedId(id), getDraftId(id)]), queryHash = hashString(PREVIEW_PROJECTION);
  return {
    query: `*[_id in $__ids_${queryHash}]${PREVIEW_PROJECTION}`,
    params: {
      [`__ids_${queryHash}`]: allIds
    }
  };
}
const BATCH_DEBOUNCE_TIME$1 = 50, isSetEqual$1 = (a, b) => a.size === b.size && Array.from(a).every((i) => b.has(i)), subscribeToStateAndFetchBatches$1 = ({
  state,
  instance,
  key: { projectId, dataset }
}) => state.observable.pipe(
  map(({ subscriptions }) => new Set(Object.keys(subscriptions))),
  distinctUntilChanged(isSetEqual$1),
  debounceTime(BATCH_DEBOUNCE_TIME$1),
  startWith$1(/* @__PURE__ */ new Set()),
  pairwise$1(),
  tap$1(([prevIds, currIds]) => {
    const newIds = [...currIds].filter((element) => !prevIds.has(element));
    state.set("updatingPending", (prev) => {
      const pendingValues = newIds.reduce((acc, id) => {
        const prevValue = prev.values[id], value = prevValue?.data ? prevValue.data : null;
        return acc[id] = { data: value, isPending: !0 }, acc;
      }, {});
      return { values: { ...prev.values, ...pendingValues } };
    });
  }),
  map(([, ids]) => ids),
  distinctUntilChanged(isSetEqual$1)
).pipe(
  switchMap((ids) => {
    if (!ids.size) return EMPTY;
    const { query, params } = createPreviewQuery(ids), controller = new AbortController();
    return new Observable((observer) => {
      const { getCurrent, observable } = getQueryState(instance, {
        query,
        params,
        tag: PREVIEW_TAG,
        perspective: PREVIEW_PERSPECTIVE,
        projectId,
        dataset
      }), subscription = defer(() => getCurrent() === void 0 ? from(
        resolveQuery(instance, {
          query,
          params,
          tag: PREVIEW_TAG,
          perspective: PREVIEW_PERSPECTIVE,
          signal: controller.signal,
          projectId,
          dataset
        })
      ).pipe(switchMap(() => observable)) : observable).pipe(filter((result) => result !== void 0)).subscribe(observer);
      return () => {
        controller.signal.aborted || controller.abort(), subscription.unsubscribe();
      };
    }).pipe(map((data) => ({ data, ids })));
  }),
  map(({ ids, data }) => ({
    values: processPreviewQuery({
      projectId,
      dataset,
      ids,
      results: data
    })
  }))
).subscribe({
  next: ({ values }) => {
    state.set("updateResult", (prev) => ({ values: { ...prev.values, ...values } }));
  }
}), previewStore = {
  name: "Preview",
  getInitialState() {
    return {
      subscriptions: {},
      values: {}
    };
  },
  initialize: (context) => {
    const subscription = subscribeToStateAndFetchBatches$1(context);
    return () => subscription.unsubscribe;
  }
};
function getPreviewState(...args) {
  return _getPreviewState(...args);
}
const _getPreviewState = bindActionByDataset(
  previewStore,
  createStateSourceAction({
    selector: ({ state }, docHandle) => state.values[docHandle.documentId] ?? STABLE_EMPTY_PREVIEW,
    onSubscribe: ({ state }, docHandle) => {
      const subscriptionId = insecureRandomId(), documentId = getPublishedId(docHandle.documentId);
      return state.set("addSubscription", (prev) => ({
        subscriptions: {
          ...prev.subscriptions,
          [documentId]: {
            ...prev.subscriptions[documentId],
            [subscriptionId]: !0
          }
        }
      })), () => {
        state.set("removeSubscription", (prev) => {
          const documentSubscriptions = omit(prev.subscriptions[documentId], subscriptionId), hasSubscribers = !!Object.keys(documentSubscriptions).length, prevValue = prev.values[documentId], previewValue = prevValue?.data ? prevValue.data : null;
          return {
            subscriptions: hasSubscribers ? { ...prev.subscriptions, [documentId]: documentSubscriptions } : omit(prev.subscriptions, documentId),
            values: hasSubscribers ? prev.values : { ...prev.values, [documentId]: { data: previewValue, isPending: !1 } }
          };
        });
      };
    }
  })
), resolvePreview = bindActionByDataset(
  previewStore,
  ({ instance }, docHandle) => firstValueFrom(getPreviewState(instance, docHandle).observable.pipe(filter((i) => !!i.data)))
), PROJECTION_TAG = "projection", PROJECTION_STATE_CLEAR_DELAY = 1e3, STABLE_EMPTY_PROJECTION = {
  data: null,
  isPending: !1
};
function validateProjection(projection) {
  if (!projection.startsWith("{") || !projection.endsWith("}"))
    throw new Error(
      `Invalid projection format: "${projection}". Projections must be enclosed in curly braces, e.g. "{title, 'author': author.name}"`
    );
  return projection;
}
function createProjectionQuery(documentIds, documentProjections) {
  const projections = Array.from(documentIds).flatMap((id) => {
    const projectionsForDoc = documentProjections[id];
    return projectionsForDoc ? Object.entries(projectionsForDoc).map(([projectionHash, projection]) => ({
      documentId: id,
      projection: validateProjection(projection),
      projectionHash
    })) : [];
  }).reduce((acc, { documentId, projection, projectionHash }) => {
    const obj = acc[projectionHash] ?? { documentIds: /* @__PURE__ */ new Set(), projection };
    return obj.documentIds.add(documentId), acc[projectionHash] = obj, acc;
  }, {}), query = `[${Object.entries(projections).map(([projectionHash, { projection }]) => `...*[_id in $__ids_${projectionHash}]{_id,_type,_updatedAt,"__projectionHash":"${projectionHash}","result":{...${projection}}}`).join(",")}]`, params = Object.fromEntries(
    Object.entries(projections).map(([projectionHash, value]) => {
      const idsInProjection = Array.from(value.documentIds).flatMap((id) => DocumentId(id));
      return [`__ids_${projectionHash}`, Array.from(idsInProjection)];
    })
  );
  return { query, params };
}
function processProjectionQuery({
  ids,
  results,
  documentStatuses
}) {
  const groupedResults = {};
  for (const result of results) {
    const originalId = getPublishedId(result._id), hash = result.__projectionHash;
    ids.has(originalId) && (groupedResults[originalId] || (groupedResults[originalId] = {}), groupedResults[originalId][hash] || (groupedResults[originalId][hash] = void 0), groupedResults[originalId][hash] = result);
  }
  const finalValues = {};
  for (const originalId of ids) {
    finalValues[originalId] = {};
    const projectionsForDoc = groupedResults[originalId];
    if (projectionsForDoc)
      for (const hash in projectionsForDoc) {
        const projectionResultData = projectionsForDoc[hash]?.result;
        if (!projectionResultData) {
          finalValues[originalId][hash] = { data: null, isPending: !1 };
          continue;
        }
        const statusFromStore = documentStatuses?.[originalId];
        finalValues[originalId][hash] = {
          data: { ...projectionResultData, _status: statusFromStore },
          isPending: !1
        };
      }
  }
  return finalValues;
}
function buildStatusQueryIds(documentIds, perspective) {
  const ids = [], releaseName = isReleasePerspective(perspective) ? perspective.releaseName : null;
  for (const id of documentIds) {
    const publishedId = getPublishedId$2(DocumentId(id)), draftId = getDraftId$1(publishedId);
    ids.push(draftId, publishedId), releaseName && ids.push(getVersionId(publishedId, releaseName));
  }
  return ids;
}
function processStatusQueryResults(results) {
  const documentStatuses = {};
  for (const result of results) {
    const id = DocumentId(result._id), updatedAt = result._updatedAt, publishedId = getPublishedId$2(id), statusData = documentStatuses[publishedId] ?? {};
    isDraftId(id) ? statusData.lastEditedDraftAt = updatedAt : isVersionId(id) ? statusData.lastEditedVersionAt = updatedAt : isPublishedId(id) && (statusData.lastEditedPublishedAt = updatedAt), documentStatuses[publishedId] = statusData;
  }
  return documentStatuses;
}
const BATCH_DEBOUNCE_TIME = 50, isSetEqual = (a, b) => a.size === b.size && Array.from(a).every((i) => b.has(i)), subscribeToStateAndFetchBatches = ({
  state,
  instance,
  key: { source, perspective }
}) => {
  const documentProjections$ = state.observable.pipe(
    map((s) => s.documentProjections),
    distinctUntilChanged(isEqual)
  ), activeDocumentIds$ = state.observable.pipe(
    map(({ subscriptions }) => new Set(Object.keys(subscriptions))),
    distinctUntilChanged(isSetEqual)
  ), pendingUpdateSubscription = activeDocumentIds$.pipe(
    debounceTime(BATCH_DEBOUNCE_TIME),
    startWith$1(/* @__PURE__ */ new Set()),
    pairwise$1(),
    tap$1(([prevIds, currIds]) => {
      const newIds = [...currIds].filter((id) => !prevIds.has(id));
      newIds.length !== 0 && state.set("updatingPending", (prev) => {
        const nextValues = { ...prev.values };
        for (const id of newIds) {
          const projectionsForDoc = prev.documentProjections[id];
          if (!projectionsForDoc) continue;
          const updatedValuesForDoc = { ...prev.values[id] ?? {} };
          for (const hash in projectionsForDoc) {
            const currentValue = updatedValuesForDoc[hash];
            updatedValuesForDoc[hash] = {
              data: currentValue?.data ?? null,
              isPending: !0
            };
          }
          nextValues[id] = updatedValuesForDoc;
        }
        return { values: nextValues };
      });
    })
  ).subscribe(), queryExecutionSubscription = combineLatest([activeDocumentIds$, documentProjections$]).pipe(
    debounceTime(BATCH_DEBOUNCE_TIME),
    distinctUntilChanged(isEqual)
  ).pipe(
    switchMap(([ids, documentProjections]) => {
      if (!ids.size) return EMPTY;
      const { query, params } = createProjectionQuery(ids, documentProjections), controller = new AbortController(), statusQueryIds = buildStatusQueryIds(ids, perspective), statusQuery = "*[_id in $statusIds]{_id, _updatedAt}", statusParams = { statusIds: statusQueryIds }, projectionQuery$ = new Observable((observer) => {
        const { getCurrent, observable } = getQueryState(instance, {
          query,
          params,
          tag: PROJECTION_TAG,
          perspective,
          // temporary guard here until we're ready for everything to be queried via global API
          ...source && !isDatasetSource(source) ? { source } : {}
        }), subscription = defer(() => getCurrent() === void 0 ? from(
          resolveQuery(instance, {
            query,
            params,
            tag: PROJECTION_TAG,
            signal: controller.signal,
            perspective,
            // temporary guard here until we're ready for everything to be queried via global API in v3
            ...source && !isDatasetSource(source) ? { source } : {}
          })
        ).pipe(switchMap(() => observable)) : observable).pipe(filter((result) => result !== void 0)).subscribe(observer);
        return () => {
          controller.signal.aborted || controller.abort(), subscription.unsubscribe();
        };
      }), statusQuery$ = new Observable((observer) => {
        const { getCurrent, observable } = getQueryState(instance, {
          query: statusQuery,
          params: statusParams,
          tag: PROJECTION_TAG,
          perspective: "raw",
          // temporary guard here until we're ready for everything to be queried via global API
          ...source && !isDatasetSource(source) ? { source } : {}
        }), subscription = defer(() => getCurrent() === void 0 ? from(
          resolveQuery(instance, {
            query: statusQuery,
            params: statusParams,
            tag: PROJECTION_TAG,
            signal: controller.signal,
            perspective: "raw",
            // temporary guard here until we're ready for everything to be queried via global API
            ...source && !isDatasetSource(source) ? { source } : {}
          })
        ).pipe(switchMap(() => observable)) : observable).pipe(filter((result) => result !== void 0)).subscribe(observer);
        return () => {
          subscription.unsubscribe();
        };
      });
      return combineLatest([projectionQuery$, statusQuery$]).pipe(
        filter(
          (pair) => pair[0] !== void 0 && pair[1] !== void 0
        ),
        map(([projection, status]) => ({
          data: projection,
          ids,
          statusResults: status
        }))
      );
    }),
    map(({ ids, data, statusResults }) => {
      const documentStatuses = processStatusQueryResults(statusResults);
      state.set("updateStatuses", (prev) => ({
        documentStatuses: {
          ...prev.documentStatuses,
          ...documentStatuses
        }
      }));
      const currentState = state.get();
      return processProjectionQuery({
        ids,
        results: data,
        documentStatuses: currentState.documentStatuses
      });
    })
  ).subscribe({
    next: (processedValues) => {
      state.set("updateResult", (prev) => {
        const nextValues = { ...prev.values };
        for (const docId in processedValues)
          processedValues[docId] && (nextValues[docId] = {
            ...prev.values[docId] ?? {},
            ...processedValues[docId]
          });
        return { values: nextValues };
      });
    },
    error: (err) => {
      console.error("Error fetching projection batches:", err);
    }
  });
  return new Subscription(() => {
    pendingUpdateSubscription.unsubscribe(), queryExecutionSubscription.unsubscribe();
  });
}, projectionStore = {
  name: "Projection",
  getInitialState() {
    return {
      values: {},
      documentProjections: {},
      subscriptions: {},
      documentStatuses: {}
    };
  },
  initialize(context) {
    const batchSubscription = subscribeToStateAndFetchBatches(context);
    return () => batchSubscription.unsubscribe();
  }
};
function getProjectionState(...args) {
  return _getProjectionState(...args);
}
const _getProjectionState = bindActionBySourceAndPerspective(
  projectionStore,
  createStateSourceAction({
    selector: ({ state }, options) => {
      const documentId = getPublishedId$2(DocumentId(options.documentId)), projectionHash = hashString(options.projection);
      return state.values[documentId]?.[projectionHash] ?? STABLE_EMPTY_PROJECTION;
    },
    onSubscribe: ({ state }, options) => {
      const { projection, ...docHandle } = options, subscriptionId = insecureRandomId(), documentId = getPublishedId$2(DocumentId(docHandle.documentId)), validProjection = validateProjection(projection), projectionHash = hashString(validProjection);
      return state.set("addSubscription", (prev) => ({
        documentProjections: {
          ...prev.documentProjections,
          [documentId]: {
            ...prev.documentProjections[documentId],
            [projectionHash]: validProjection
          }
        },
        subscriptions: {
          ...prev.subscriptions,
          [documentId]: {
            ...prev.subscriptions[documentId],
            [projectionHash]: {
              ...prev.subscriptions[documentId]?.[projectionHash],
              [subscriptionId]: !0
            }
          }
        }
      })), () => {
        setTimeout(() => {
          state.set("removeSubscription", (prev) => {
            const documentSubscriptionsForHash = omit(
              prev.subscriptions[documentId]?.[projectionHash],
              subscriptionId
            ), hasSubscribersForProjection = !!Object.keys(documentSubscriptionsForHash).length, nextSubscriptions = { ...prev.subscriptions }, nextDocumentProjections = { ...prev.documentProjections }, nextValues = { ...prev.values };
            if (hasSubscribersForProjection)
              nextSubscriptions[documentId] && (nextSubscriptions[documentId][projectionHash] = documentSubscriptionsForHash);
            else {
              delete nextSubscriptions[documentId][projectionHash], delete nextDocumentProjections[documentId][projectionHash];
              const currentProjectionValue = prev.values[documentId]?.[projectionHash];
              currentProjectionValue && nextValues[documentId] && (nextValues[documentId][projectionHash] = {
                data: currentProjectionValue.data,
                isPending: !1
              });
            }
            return Object.values(
              nextSubscriptions[documentId] ?? {}
            ).some((subs) => Object.keys(subs).length > 0) || (delete nextSubscriptions[documentId], delete nextDocumentProjections[documentId]), {
              subscriptions: nextSubscriptions,
              documentProjections: nextDocumentProjections,
              values: nextValues
            };
          });
        }, PROJECTION_STATE_CLEAR_DELAY);
      };
    }
  })
);
function resolveProjection(...args) {
  return _resolveProjection(...args);
}
const _resolveProjection = bindActionBySourceAndPerspective(
  projectionStore,
  ({ instance }, options) => firstValueFrom(
    getProjectionState(instance, options).observable.pipe(
      filter((state) => !!state?.data)
    )
  )
), API_VERSION = "v2025-02-19", projects = createFetcherStore({
  name: "Projects",
  getKey: (_instance, options) => {
    const orgKey = options?.organizationId ? `:org:${options.organizationId}` : "", membersKey = options?.includeMembers === !1 ? ":no-members" : "";
    return `projects${orgKey}${membersKey}`;
  },
  fetcher: (instance) => (options) => getClientState(instance, {
    apiVersion: API_VERSION,
    scope: "global",
    requestTagPrefix: "sanity.sdk.projects"
  }).observable.pipe(
    switchMap((client) => {
      const organizationId = options?.organizationId;
      return client.observable.projects.list({
        // client method has a type that expects false | undefined
        includeMembers: options?.includeMembers ? void 0 : !1,
        organizationId
      });
    })
  )
}), getProjectsState = projects.getState, resolveProjects = projects.resolveState, TOKEN_REGEX = /(?:[^\s"]+|"[^"]*")+/g;
function isNegationToken(token) {
  return typeof token < "u" && token.trim().startsWith("-");
}
function isPrefixToken(token) {
  return typeof token < "u" && token.trim().endsWith("*");
}
function isExactMatchToken(token) {
  return !!token && token.length >= 2 && token.startsWith('"') && token.endsWith('"');
}
function createGroqSearchFilter(query) {
  const trimmedQuery = query.trim();
  if (!trimmedQuery)
    return "";
  const tokens = trimmedQuery.match(TOKEN_REGEX) ?? [], reversedIndex = [...tokens].reverse().findIndex(
    (token) => !isNegationToken(token) && !isExactMatchToken(token)
  ), finalIncrementalTokenIndex = reversedIndex === -1 ? -1 : tokens.length - 1 - reversedIndex, finalIncrementalToken = tokens[finalIncrementalTokenIndex], processedTokens = [...tokens];
  return finalIncrementalToken !== void 0 && !isPrefixToken(finalIncrementalToken) && processedTokens.splice(
    finalIncrementalTokenIndex,
    1,
    `${finalIncrementalToken}*`
  ), `[@] match text::query("${processedTokens.join(" ").replace(/"/g, '\\"')}")`;
}
function defineIntent(intent) {
  if (!intent.id)
    throw new Error("Intent must have an id");
  if (!intent.action)
    throw new Error("Intent must have an action");
  if (!intent.title)
    throw new Error("Intent must have a title");
  if (!Array.isArray(intent.filters))
    throw new Error("Intent must have a filters array");
  if (intent.filters.length === 0)
    throw new Error(
      "Intent must have at least one filter. If you want to match everything, use {types: ['*']}"
    );
  return intent.filters.forEach((filter2, index) => {
    validateFilter(filter2, index);
  }), intent;
}
function validateFilter(filter2, index) {
  const filterContext = `Filter at index ${index}`;
  if (!filter2 || typeof filter2 != "object")
    throw new Error(`${filterContext} must be an object`);
  if (filter2.types === void 0)
    throw new Error(
      `${filterContext} must have a types property. Use ['*'] to match all document types.`
    );
  if (filter2.projectId !== void 0) {
    if (typeof filter2.projectId != "string")
      throw new Error(`${filterContext}: projectId must be a string`);
    if (filter2.projectId.trim() === "")
      throw new Error(`${filterContext}: projectId cannot be empty`);
  }
  if (filter2.dataset !== void 0) {
    if (typeof filter2.dataset != "string")
      throw new Error(`${filterContext}: dataset must be a string`);
    if (filter2.dataset.trim() === "")
      throw new Error(`${filterContext}: dataset cannot be empty`);
    if (filter2.projectId === void 0)
      throw new Error(`${filterContext}: dataset cannot be specified without projectId`);
  }
  if (!Array.isArray(filter2.types))
    throw new Error(`${filterContext}: types must be an array`);
  if (filter2.types.length === 0)
    throw new Error(`${filterContext}: types array cannot be empty`);
  if (filter2.types.forEach((type, typeIndex) => {
    if (typeof type != "string")
      throw new Error(`${filterContext}: types[${typeIndex}] must be a string`);
    if (type.trim() === "")
      throw new Error(`${filterContext}: types[${typeIndex}] cannot be empty`);
  }), filter2.types.includes("*") && filter2.types.length > 1)
    throw new Error(
      `${filterContext}: when using wildcard '*', it must be the only type in the array`
    );
}
function getCorsErrorProjectId(error) {
  if (!(error instanceof CorsOriginError)) return null;
  const projMatch = (error.message || "").match(/manage\/project\/([^/?#]+)/);
  return projMatch ? projMatch[1] : null;
}
var version = "2.8.0";
const CORE_SDK_VERSION = getEnv("PKG_VERSION") || `${version}-development`;
export {
  AuthStateType,
  CORE_SDK_VERSION,
  agentGenerate,
  agentPatch,
  agentPrompt,
  agentTransform,
  agentTranslate,
  applyDocumentActions,
  configureLogging,
  createDatasetHandle,
  createDocument,
  createDocumentHandle,
  createDocumentTypeHandle,
  createGroqSearchFilter,
  createProjectHandle,
  createSanityInstance,
  defineIntent,
  deleteDocument,
  destroyController,
  discardDocument,
  editDocument,
  getActiveReleasesState,
  getAuthState,
  getClient,
  getClientErrorApiBody,
  getClientErrorApiDescription,
  getClientErrorApiType,
  getClientState,
  getCorsErrorProjectId,
  getCurrentUserState,
  getDashboardOrganizationId$1 as getDashboardOrganizationId,
  getDatasetsState,
  getDocumentState,
  getDocumentSyncStatus,
  getFavoritesState,
  getIndexForKey2 as getIndexForKey,
  getIsInDashboardState,
  getLoginUrlState,
  getNodeState,
  getOrCreateChannel,
  getOrCreateController,
  getOrCreateNode,
  getPathDepth,
  getPermissionsState,
  getPerspectiveState,
  getPresence,
  getPreviewState,
  getProjectState,
  getProjectionState,
  getProjectsState,
  getQueryKey,
  getQueryState,
  getTokenState,
  getUserState,
  getUsersKey,
  getUsersState,
  handleAuthCallback,
  isCanvasSource,
  isDatasetSource,
  isMediaLibrarySource,
  isProjectUserNotFoundClientError,
  isStudioConfig,
  joinPaths,
  jsonMatch2 as jsonMatch,
  loadMoreUsers,
  logout,
  observeOrganizationVerificationState,
  parseQueryKey,
  parseUsersKey,
  publishDocument,
  releaseChannel,
  releaseNode,
  resolveDatasets,
  resolveDocument,
  resolveFavoritesState,
  resolvePermissions,
  resolvePreview,
  resolveProject,
  resolveProjection,
  resolveProjects,
  resolveQuery,
  resolveUser,
  resolveUsers,
  setAuthToken,
  slicePath2 as slicePath,
  stringifyPath2 as stringifyPath,
  subscribeDocumentEvents,
  unpublishDocument
};
